from os import path
from glob import glob
from pyppl import Proc, Box
#from .utils import plot, txt, dirnamePattern
from . import params, rimport
from .utils import dirpat2name

"""
@name:
	pEXPRdir2Matrix
@description:
	Convert expression files to expression matrix
	File names will be used as sample names (colnames)
	Each gene and its expression per line.
	Suppose each expression file has the same rownames and in the same order.
@input:
	`indir:file`:  the directory containing the expression files, could be gzipped
@output:
	`outfile:file`: the expression matrix file
	`outdir:dir`:   the directory containing expr file and plots
@args:
	`pattern` : The pattern to filter files. Default `'*'`
	`namefunc`: Transform filename (no extension) as column name. Default: "function(fn) fn"
	`header`  : Whether each expression file contains header. Default: `False`
	`exrows`  : Rows to be excluded, regular expression applied. Default: `["^Sample", "^Composite", "^__"]`
	`boxplot` : Whether to plot a boxplot. Default: False
	`heatmap` : Whether to plot a heatmap. Default: False
	`histplot`: Whether to plot a histgram. Default: False
	`devpars` : Parameters for png. Default: `{'res': 300, 'width': 2000, 'height': 2000}`
	`boxplotggs`: The ggplot parameters for boxplot. Default: `['r:ylab("Expression")']`
		- See ggplot2 documentation.
	`heatmapggs`: The ggplot parameters for heatmap. Default: `['r:theme(axis.text.y = element_blank())']`
	`histplotggs`: The ggplot parameters for histgram. Default: `['r:labs(x = "Expression", y = "# Samples")']`
"""
pEXPRdir2Matrix        = Proc(desc = 'Merge expression files to a matrix.')
pEXPRdir2Matrix.input  = "indir:file"
pEXPRdir2Matrix.output = [
	"outfile:file:{{in.indir, args.pattern | dirpat2name}}.dir/{{in.indir, args.pattern | dirpat2name}}.expr.txt",
	"outdir:dir:{{in.indir, args.pattern | dirpat2name}}.dir"
]
pEXPRdir2Matrix.lang           = params.Rscript.value
pEXPRdir2Matrix.args.pattern   = '*'
pEXPRdir2Matrix.args.fn2sample = 'function(fn) fn'
pEXPRdir2Matrix.args.exrows    = ["^Sample", "^Composite", "^__"]
pEXPRdir2Matrix.args.hmrows    = 500
pEXPRdir2Matrix.args.plot = Box(
	boxplot = False,
	heatmap = False,
	histogram = False
)
pEXPRdir2Matrix.args.ggs = Box(
	boxplot = Box(ylab = {0: "Expression"}),
	heatmap = Box(theme = {'axis.text.y': 'r:element_blank()'}),
	histogram = Box(labs = {'x': 'Expression', 'y': '# Genes'})
)
pEXPRdir2Matrix.envs.dirpat2name = dirpat2name
pEXPRdir2Matrix.envs.rimport = rimport
pEXPRdir2Matrix.args.devpars = Box(res = 300, width = 2000, height = 2000)
pEXPRdir2Matrix.script       = "file:scripts/rnaseq/pEXPRdir2Matrix.r"

"""
@name:
	pBatchEffect
@description:
	Remove batch effect with sva-combat.
@input:
	`expr:file`:  The expression file, generated by pEXPRdir2Matrix
	`batch:file`: The batch file defines samples and batches.
@output:
	`outfile:file`: the expression matrix file
	`outdir:dir`:   the directory containing expr file and plots
@args:
	`tool`    : The tool used to remove batch effect. Default `'combat'`
	`hmrows`  : How many rows to be used to plot heatmap
	`plot`: Whether to plot
		- `boxplot`   : Whether to plot a boxplot. Default: False
		- `heatmap`   : Whether to plot a heatmap. Default: False
		- `histogram` : Whether to plot a histgram. Default: False
	`devpars`    : Parameters for png. Default: `{'res': 300, 'width': 2000, 'height': 2000}`
	`ggs`: The ggplot parameters
		- `boxplot`  : The ggplot parameters for boxplot. Default: `Box(ylab = {0: "Log2 Intensity"})`
		- `heatmap`  : The ggplot parameters for heatmap. Default: `Box(theme = {'axis.text.y': 'r:element_blank()'})`
		- `histogram`: The ggplot parameters for histgram. Default: `Box(labs = {'x': "Log2 Intensity", "y": "Density"})`
"""
pBatchEffect              = Proc(desc = 'Try to remove batch effect of expression data.')
pBatchEffect.input        = "expr:file, batch:file"
pBatchEffect.output       = "outfile:file:{{in.expr | fn2}}/{{in.expr | fn2}}.expr.txt, outdir:dir:{{in.expr | fn2}}"
pBatchEffect.args.tool    = 'combat'
pBatchEffect.args.plot    = Box(boxplot = False, heatmap = False, histogram = False)
pBatchEffect.args.hmrows  = 500
pBatchEffect.args.devpars = Box(res = 300, width = 2000, height = 2000)
pBatchEffect.args.ggs     = Box(
	boxplot   = Box(ylab = {0: "Expression"}),
	heatmap   = Box(theme = {'axis.text.y': 'r:element_blank()'}),
	histogram = Box(labs  = {'x': "Expression", "y": "Frequency"})
)
pBatchEffect.envs.rimport = rimport
pBatchEffect.lang   = params.Rscript.value
pBatchEffect.script = "file:scripts/rnaseq/pBatchEffect.r"

"""
@name:
	pRawCounts2
@description:
	Convert raw counts to another unit
@input:
	`infile:file`: the expression matrix
		- rows are genes, columns are samples
@output:
	`outfile:file`: the converted expression matrix
@args:
	`transpose`: transpose the input matrix? default: False
	`log2`:      whether to take log2? default: False
	`unit`:      convert to which unit? default: cpm (or rpkm, tmm)
	`header`:    whether input file has header? default: True
	`rownames`:  the index of the column as rownames. default: 1
	`glenfile`:  the gene length file, for RPKM
		- no head, row names are genes, have to be exact the same order and length as the rownames of infile
	`boxplot` : Whether to plot a boxplot. Default: False
	`heatmap` : Whether to plot a heatmap. Default: False
	`histplot`: Whether to plot a histgram. Default: False
	`devpars` : Parameters for png. Default: `{'res': 300, 'width': 2000, 'height': 2000}`
	`boxplotggs`: The ggplot parameters for boxplot. Default: `['r:ylab("Expression")']`
		- See ggplot2 documentation.
	`heatmapggs`: The ggplot parameters for heatmap. Default: `['r:theme(axis.text.y = element_blank())']`
	`histplotggs`: The ggplot parameters for histgram. Default: `['r:labs(x = "Expression", y = "# Samples")']`
@requires:
	[edgeR](https://bioconductor.org/packages/release/bioc/html/edger.html) if cpm or rpkm is chosen
	[coseq](https://rdrr.io/rforge/coseq/man/transform_RNAseq.html) if tmm is chosen
"""
pRawCounts2              = Proc(desc = 'Convert raw counts to another unit.')
pRawCounts2.input        = "infile:file"
pRawCounts2.output       = "outfile:file:{{in.infile | fn2}}.dir/{{in.infile | fn2}}.expr.txt, outdir:dir:{{in.infile | fn2}}.dir"
pRawCounts2.args.unit    = 'cpm'
pRawCounts2.args.refgene = params.refgene.value
pRawCounts2.args.hmrows  = 500
pRawCounts2.args.plot    = Box(
	boxplot   = False,
	heatmap   = False,
	histogram = False
)
pRawCounts2.args.ggs = Box(
	boxplot   = Box(ylab = {0: 'Expression'}),
	heatmap   = Box(theme = {'axis.text.y': 'r:element_blank()'}),
	histogram = Box(labs = {'x': 'Expression', 'y': '# Genes'})
)
pRawCounts2.args.devpars = Box(res = 300, width = 2000, height = 2000)
pRawCounts2.envs.rimport = rimport
pRawCounts2.lang         = params.Rscript.value
pRawCounts2.script       = "file:scripts/rnaseq/pRawCounts2.r"

"""
@name:
	p2RawCounts
@description:
	Convert gene expression to raw counts.
@input:
	`infile:file`: The expression matrix file
@output:
	`outfile:file`: The output file
	`outdir:dir`:   The output directory, may contain the figures.
@args:
	`unit`: The unit of input gene expression. Default: fpkm
		- Could also be rpkm, tpm
	`nreads`     : Total reads approximately. Default: 30, 000, 000
	`refgene`    : The refgene file for gene length.
	`boxplot`    : Whether to plot boxplot after transformation. Default: False
	`heatmap`    : Whether to plot heatmap after transformation. Default: False
	`heatmapn`   : How many genes used to plot heatmap. Default: 500
	`histplot`   : Whether to plot histgram after transformation. Default: False
	`devpars`    : The device parameters for plotting. Default: `{'res': 300, 'width': 2000, 'height': 2000}`
	`boxplotggs` : The ggplot statement for boxplot.
	`heatmapggs` : The ggplot statement for heatmap.
	`histplotggs`: The ggplot statement for histgram.
"""
p2RawCounts              = Proc(desc = 'Convert normalized expression back to raw counts.')
p2RawCounts.input        = 'infile:file'
p2RawCounts.output       = 'outfile:file:{{in.infile | fn2}}.dir/{{in.infile | fn2}}.counts.txt, outdir:dir:{{in.infile | fn2}}.dir'
p2RawCounts.args.unit    = 'fpkm'
p2RawCounts.args.nreads  = 30000000
p2RawCounts.args.refgene = params.refgene.value
p2RawCounts.args.hmrows  = 500
p2RawCounts.args.plot    = Box(
	boxplot   = False,
	heatmap   = False,
	histogram = False
)
p2RawCounts.args.ggs = Box(
	boxplot   = Box(ylab = {0: 'Expression'}),
	heatmap   = Box(theme = {'axis.text.y': 'r:element_blank()'}),
	histogram = Box(labs = {'x': 'Expression', 'y': '# Genes'})
)
p2RawCounts.args.devpars = Box(res = 300, width = 2000, height = 2000)
p2RawCounts.envs.rimport = rimport
p2RawCounts.lang         = params.Rscript.value
p2RawCounts.script       = "file:scripts/rnaseq/p2RawCounts.r"


"""
@name:
	pRNAseqDEG
@description:
	Detect DEGs for RNA-seq data
@input:
	`efile:file`: The expression matrix
	`gfile:file`: The group information
		- Like:
		```
		Sample1	Group1
		Sample2	Group1
		Sample3	Group1
		Sample4	group2
		Sample5	group2
		Sample6	group2
		```
@output:
	`outfile:file`: The DEG list
	`outdir:file`:  The output directory containing deg list and plots
@args:
	`tool`      : the tool used to detect DEGs. Default: 'edger' (deseq2)
	`filter`    : filter out low count records. Default: `"1,2"` (At least 2 samples have at least 2 reads)
	`mdsplot`   : whether to plot the MDS plot, default : True
	`volplot`   : whether to plot the volcano plot, default : True
	`maplot`    : whether to plot MA plots within each group, default : False
	`heatmap`   : whether to plot the heatmap using DEGs. Default : False
	`heatmapn`  : How many genes to be used for heatmap. If `heatmapn`, the number will be `heatmapn * # DEGs`. Default: 100
	`heatmapggs`: The ggplots options for heatmap. Default : []
	`maplotggs` : The ggplots options for maplot. Default : []
	`volplotggs`: The ggplots options for volplot. Default : []
	`devpars`   : Parameters for png. Default: `{'res': 300, 'width': 2000, 'height': 2000}`
"""
pRNAseqDEG        = Proc(desc = 'Detect DEGs by RNA-seq data.')
pRNAseqDEG.input  = "efile:file, gfile:file"
pRNAseqDEG.output = [
	"outfile:file:{{in.efile | fn2}}-{{in.gfile | fn2}}.DEGs/{{in.efile | fn2}}-{{in.gfile | fn2}}.degs.txt",
	"outdir:dir:{{in.efile | fn2}}-{{in.gfile | fn2}}.DEGs"
]
pRNAseqDEG.args.tool   = 'edger' # deseq2
pRNAseqDEG.args.filter = '1,2'
pRNAseqDEG.args.pval   = 0.05
pRNAseqDEG.args.hmrows = 100
pRNAseqDEG.args.plot = Box(
	mdsplot = True,
	volplot = True,
	maplot  = False,
	heatmap = False
)
pRNAseqDEG.args.ggs = Box(
	maplot  = Box(),
	heatmap = Box(theme = {'axis.text.y': 'r:element_blank()'}),
	volplot = Box()
)

pRNAseqDEG.args.devpars = Box(res = 300, width = 2000, height = 2000)
pRNAseqDEG.envs.rimport = rimport
pRNAseqDEG.lang         = params.Rscript.value
pRNAseqDEG.script       = "file:scripts/rnaseq/pRNAseqDEG.r"


pCoexp             = Proc(desc = "Get co-expression of gene pairs in the expression matrix.")
pCoexp.input       = "infile:file"
pCoexp.output      = "outfile:file:{{in.infile | fn}}.coexp, outpval:file:{{in.infile | fn}}.pval"
pCoexp.args.method = 'pearson'
pCoexp.args.pval   = False
pCoexp.lang        = params.Rscript.value
pCoexp.script      = "file:scripts/rnaseq/pCoexp.r"
