<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Processes - bioprocs</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../css/highlight.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Processes";
    var mkdocs_page_input_path = "processes.md";
    var mkdocs_page_url = "/processes/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js"></script>
  <script src="../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../js/highlight.pack.js"></script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> bioprocs</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Introduction</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">Processes</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#algorithm">algorithm</a></li>
    

    <li class="toctree-l2"><a href="#bed">bed</a></li>
    

    <li class="toctree-l2"><a href="#bedtools">bedtools</a></li>
    

    <li class="toctree-l2"><a href="#chipseq">chipseq</a></li>
    

    <li class="toctree-l2"><a href="#cluster">cluster</a></li>
    

    <li class="toctree-l2"><a href="#cnvkit">cnvkit</a></li>
    

    <li class="toctree-l2"><a href="#common">common</a></li>
    

    <li class="toctree-l2"><a href="#eqtl">eqtl</a></li>
    

    <li class="toctree-l2"><a href="#fastx">fastx</a></li>
    

    <li class="toctree-l2"><a href="#gatk">gatk</a></li>
    

    <li class="toctree-l2"><a href="#gene">gene</a></li>
    

    <li class="toctree-l2"><a href="#genomeplot">genomeplot</a></li>
    

    <li class="toctree-l2"><a href="#gsea">gsea</a></li>
    

    <li class="toctree-l2"><a href="#hic">hic</a></li>
    

    <li class="toctree-l2"><a href="#marray">marray</a></li>
    

    <li class="toctree-l2"><a href="#misc">misc</a></li>
    

    <li class="toctree-l2"><a href="#pca">pca</a></li>
    

    <li class="toctree-l2"><a href="#picard">picard</a></li>
    

    <li class="toctree-l2"><a href="#plot">plot</a></li>
    

    <li class="toctree-l2"><a href="#power">power</a></li>
    

    <li class="toctree-l2"><a href="#rank">rank</a></li>
    

    <li class="toctree-l2"><a href="#resource">resource</a></li>
    

    <li class="toctree-l2"><a href="#rnaseq">rnaseq</a></li>
    

    <li class="toctree-l2"><a href="#sambam">sambam</a></li>
    

    <li class="toctree-l2"><a href="#seq">seq</a></li>
    

    <li class="toctree-l2"><a href="#snp">snp</a></li>
    

    <li class="toctree-l2"><a href="#snparray">snparray</a></li>
    

    <li class="toctree-l2"><a href="#sql">sql</a></li>
    

    <li class="toctree-l2"><a href="#stats">stats</a></li>
    

    <li class="toctree-l2"><a href="#tabix">tabix</a></li>
    

    <li class="toctree-l2"><a href="#tcga">tcga</a></li>
    

    <li class="toctree-l2"><a href="#tfbs">tfbs</a></li>
    

    <li class="toctree-l2"><a href="#tsv">tsv</a></li>
    

    <li class="toctree-l2"><a href="#utils">utils</a></li>
    

    <li class="toctree-l2"><a href="#vcf">vcf</a></li>
    

    <li class="toctree-l2"><a href="#vcfnext">vcfnext</a></li>
    

    <li class="toctree-l2"><a href="#web">web</a></li>
    

    <li class="toctree-l2"><a href="#wxs">wxs</a></li>
    

    <li class="toctree-l2"><a href="#wxsanno">wxsanno</a></li>
    

    <li class="toctree-l2"><a href="#wxscall">wxscall</a></li>
    

    <li class="toctree-l2"><a href="#wxsdown">wxsdown</a></li>
    

    <li class="toctree-l2"><a href="#wxsprep">wxsprep</a></li>
    

    <li class="toctree-l2"><a href="#wxsstat">wxsstat</a></li>
    

    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../aggregations/">Aggregations</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../binaries/">Binaries</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">bioprocs</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Processes</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h3 id="algorithm">algorithm</h3>
<div class="admonition hint">
<p class="admonition-title">pRWR</p>
<ul>
<li>
<p><strong>description</strong><br />
    Do random walk with restart (RWR)</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>Wfile:file</code>: The adjecent matrix  </li>
<li><code>Efile:file</code>: The start vector  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output of final probabilities  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>c</code>: The restart probability. Default: 0.1  </li>
<li><code>eps</code>: The convergent cutoff || R(i+1) - R(i) ||. Default: 1e-5  </li>
<li><code>niter</code>: Max iterations to stop. Default: 10000  </li>
<li><code>normW</code>: Weather to normalize W or not, default True.   <ul>
<li>Laplacian normalization is used (more to add).</li>
</ul>
</li>
<li><code>normE</code>: Weather to normalize E or not, default True.   <ul>
<li>E will be normalized as: E = E/sum(E)</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
    if normW = True, R package <code>NetPreProc</code> is required.</p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pAR</p>
<ul>
<li>
<p><strong>description</strong><br />
    Affinity Regression.
    Ref: https://www.nature.com/articles/nbt.3343
    <div class="highlight"><pre><span></span>        b           c        d          d  
    _________    _______    ____       ____
    |       |    |  W  |    |  |       |  |
  a |   D   |  b |_____|  c |Pt|  =  a |Y |   &lt;=&gt;
    |_______|               |__|       |  |
                                       |__|

kronecker(P, YtD)*vec(W) = vec(YtY)             &lt;=&gt;
X*vec(W) = vec(YtY)
WPt:
       c        d              d  
    _______    ____          _____
    |  W  |    |  |          |   |
  b |_____|  c |Pt|  ---&gt;  b |___|
                  |__|

YtDW:
WtDtY:
     b           a        d               d    
  _______    _________   ____           _____  
  |  Wt |    |       |   |  |           |   |  
c |_____|  b |   Dt  | a |Y |    ---&gt; c |___|  
             |_______|   |  |                 
                         |__|                  
</pre></div></p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>D:file</code> : The D matrix  </li>
<li><code>Pt:file</code>: The Pt matrix  </li>
<li><code>Y:file</code>: The Y matrix  <ul>
<li>All input files could be gzipped</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>W:file</code>: The interaction matrix  </li>
<li><code>outdir:dir</code>: The output directory  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>seed</code>: The seed for sampling the training set.  </li>
<li><code>tfrac</code>: The fraction of samples used for training.<br />
``</li>
</ul>
</li>
</ul>
</div>
<h3 id="bed">bed</h3>
<div class="admonition hint">
<p class="admonition-title">pBedSort</p>
<ul>
<li>
<p><strong>description</strong><br />
    Sort bed files</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used to sort the file. Default: sort (bedtools, bedops)  </li>
<li><code>bedtools</code>: The path to bedtools. Default: bedtools  </li>
<li><code>bedops_sort</code>: The path to bedops' sort-bed. Default: sort-bed  </li>
<li><code>mem</code>: The memory to use. Default: 8G  </li>
<li><code>tmpdir</code>: The tmpdir to use. Default: <code>$TMPDIR</code>  </li>
<li><code>unique</code>: Remove the dupliated records? Default: True  </li>
<li><code>params</code>: Other params for <code>tool</code>. Default: {}  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html"><code>bedtools</code></a>
    <a href="https://github.com/bedops/bedops"><code>bedops</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedCluster</p>
<ul>
<li>
<p><strong>description</strong><br />
    Assign cluster id to each record</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input bed file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used to sort the file. Default: bedtools  </li>
<li><code>bedtools</code>: The path to bedtools. Default: bedtools  </li>
<li><code>params</code>: Other params for <code>tool</code>. Default: ''  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html"><code>bedtools</code></a></p>
</li>
</ul>
</div>
<h3 id="bedtools">bedtools</h3>
<div class="admonition hint">
<p class="admonition-title">pBedGetfasta</p>
<ul>
<li>
<p><strong>description</strong><br />
<code>bedtools getfasta</code> extracts sequences from a FASTA file for each of the intervals defined in a BED file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input bed file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The generated fasta file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>ref</code>     : The fasta file  </li>
<li><code>bedtools</code>: The bedtools executable,                  default: "bedtools"  </li>
<li><code>params</code>  : Other parameters for <code>bedtools getfasta</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedClosest</p>
<ul>
<li>
<p><strong>description</strong><br />
    Similar to intersect, closest searches for overlapping features in A and B. In the event that no feature in B overlaps the current feature in A, closest will report the nearest (that is, least genomic distance from the start or end of A) feature in B. For example, one might want to find which is the closest gene to a significant GWAS polymorphism. Note that closest will report an overlapping feature as the closest that is, it does not restrict to closest non-overlapping feature. The following iconic cheatsheet summarizes the funcitonality available through the various optyions provided by the closest tool.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>afile:file</code>: The -a file  </li>
<li><code>bfile:file</code>: The -b file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bedtools</code>: The bedtools executable, default: "bedtools"  </li>
<li><code>params</code>: Other parameters for <code>bedtools closest</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedClosest2</p>
<ul>
<li>
<p><strong>description</strong><br />
    Multiple b-file version of pBedClosest</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>afile:file</code>: The -a file  </li>
<li><code>bfiles:files</code>: The -b files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bedtools</code>: The bedtools executable, default: "bedtools"  </li>
<li><code>params</code>: Other parameters for <code>bedtools closest</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedFlank</p>
<ul>
<li>
<p><strong>description</strong><br />
<code>bedtools flank</code> will create two new flanking intervals for each interval in a BED file. Note that flank will restrict the created flanking intervals to the size of the chromosome (i.e. no start &lt; 0 and no end &gt; chromosome size).</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
<li><code>gfile:file</code>: The genome size file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bin</code>: The bedtools executable, default: "bedtools"  </li>
<li><code>params</code>: Other parameters for <code>bedtools flank</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedIntersect</p>
<ul>
<li>
<p><strong>description</strong><br />
    By far, the most common question asked of two sets of genomic features is whether or not any of the features in the two sets overlap with one another. This is known as feature intersection. bedtools intersect allows one to screen for overlaps between two sets of genomic features. Moreover, it allows one to have fine control as to how the intersections are reported. bedtools intersect works with both BED/GFF/VCF and BAM files as input.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>afile:file</code> : The a file  </li>
<li><code>bfile:file</code>: The b file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bedtools</code>: The bedtools executable, default: "bedtools"  </li>
<li><code>params</code>: Other parameters for <code>bedtools intersect</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedIntersect2</p>
<ul>
<li>
<p><strong>description</strong><br />
    Multiple b-file version of pBedIntersect</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>afile:file</code> : The a file  </li>
<li><code>bfiles:files</code>: The b files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bedtools</code>: The bedtools executable, default: "bedtools"  </li>
<li><code>params</code>: Other parameters for <code>bedtools intersect</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedMakewindows</p>
<ul>
<li>
<p><strong>description</strong><br />
    Makes adjacent or sliding windows across a genome or BED file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bin</code>: The bedtools executable, default: "bedtools"  </li>
<li><code>informat</code>: The format of input file, whether is a "bed" file or "genome" size file. Default: "bed"  </li>
<li><code>params</code>: Other parameters for <code>bedtools makewindows</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedMerge</p>
<ul>
<li>
<p><strong>description</strong><br />
<code>bedtools merge</code> combines overlapping or book-ended features in an interval file into a single feature which spans all of the combined features.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bedtools</code>: The bedtools executable,               default: "bedtools"  </li>
<li><code>params</code>  : Other parameters for <code>bedtools merge</code>, default: {}  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedMerge2</p>
<ul>
<li>
<p><strong>description</strong><br />
    A multi-input file model of pBedMerge: Merge multiple input files.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infiles:files</code>: The input files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bedtools</code>: The bedtools executable,               default: "bedtools"  </li>
<li><code>params</code>  : Other parameters for <code>bedtools merge</code>, default: {}  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedMultiinter</p>
<ul>
<li>
<p><strong>description</strong><br />
    Identifies common intervals among multiple BED/GFF/VCF files.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infiles:files</code>: The input files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bin</code>: The bedtools executable, default: "bedtools"  </li>
<li><code>params</code>: Other parameters for <code>bedtools multiinter</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedRandom</p>
<ul>
<li>
<p><strong>description</strong><br />
<code>bedtools random</code> will generate a random set of intervals in BED6 format. One can specify both the number (-n) and the size (-l) of the intervals that should be generated.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>gfile:file</code>: The genome size file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bedtools</code>: The bedtools executable,    default: "bedtools"  </li>
<li><code>seed</code>    : The seed for randomization, default: None  </li>
<li><code>gsize</code>   : The chromsize file.  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedShift</p>
<ul>
<li>
<p><strong>description</strong><br />
<code>bedtools shift</code> will move each feature in a feature file by a user-defined number of bases. While something like this could be done with an awk '{OFS="\t" print $1,$2+<shift>,$3+<shift>}', bedtools shift will restrict the resizing to the size of the chromosome (i.e. no features before 0 or past the chromosome end).</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
<li><code>gfile:file</code>: The genome size file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bin</code>: The bedtools executable, default: "bedtools"  </li>
<li><code>params</code>: Other parameters for <code>bedtools shift</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedShuffle</p>
<ul>
<li>
<p><strong>description</strong><br />
<code>bedtools shuffle</code> will randomly permute the genomic locations of a feature file among a genome defined in a genome file. One can also provide an exclusions BED/GFF/VCF file that lists regions where you do not want the permuted features to be placed. For example, one might want to prevent features from being placed in known genome gaps. shuffle is useful as a null basis against which to test the significance of associations of one feature with another.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
<li><code>gfile:file</code>: The genome size file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bin</code>: The bedtools executable, default: "bedtools"  </li>
<li><code>params</code>: Other parameters for <code>bedtools shuffle</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedSubtract</p>
<ul>
<li>
<p><strong>description</strong><br />
<code>bedtools subtract</code> searches for features in B that overlap A. If an overlapping feature is found in B, the overlapping portion is removed from A and the remaining portion of A is reported. If a feature in B overlaps all of a feature in A, the A feature will not be reported.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>afile:file</code>: The a file  </li>
<li><code>bfile:file</code>: The b file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bin</code>: The bedtools executable, default: "bedtools"  </li>
<li><code>params</code>: Other parameters for <code>bedtools subtract</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedWindow</p>
<ul>
<li>
<p><strong>description</strong><br />
    Similar to <code>bedtools intersect</code>, <code>window</code> searches for overlapping features in A and B. However, window adds a specified number (1000, by default) of base pairs upstream and downstream of each feature in A. In effect, this allows features in B that are near features in A to be detected.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>afile:file</code>: The a file  </li>
<li><code>bfile:file</code>: The b file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bin</code>: The bedtools executable, default: "bedtools"  </li>
<li><code>params</code>: Other parameters for <code>bedtools window</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBedGenomecov</p>
<ul>
<li>
<p><strong>description</strong><br />
<code>bedtools genomecov</code> computes histograms (default), per-base reports (-d) and BEDGRAPH (-bg) summaries of feature coverage (e.g., aligned sequences) for a given genome.</p>
<p>NOTE: only bam file input implemented here.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The bam file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bin</code>: The bedtools executable, default: "bedtools"  </li>
<li><code>params</code>: Other parameters for <code>bedtools genomecov</code>, default: "-bg"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bedtools.readthedocs.io/en/latest/index.html">bedtools</a></p>
</li>
</ul>
</div>
<h3 id="chipseq">chipseq</h3>
<div class="admonition hint">
<p class="admonition-title">pPeakToRegPotential</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert peaks to regulatory potential score for each gene
    The formula is:
    <code>-(0.5 + 4*di/d0)
    PC = sum (pi * e                  )</code>
    Ref: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4489297/</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>peakfile:file</code>: The BED/peak file for peaks  </li>
<li><code>genefile:file</code>: The BED file for gene coordinates  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The regulatory potential file for each gene  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>intensity</code>: <code>pi</code> in the formula. Boolean value, whether use the peak intensity or not, default: <code>True</code>,  </li>
<li><code>geneformat</code>: The format for <code>genefile</code>, default: <code>ucsc+gz</code>. It could be:  <ul>
<li>ucsc or ucsc+gz: typically, you can download from http://hgdownload.cse.ucsc.edu/goldenPath/hg38/database/refGene.txt.gz</li>
<li>bed or bed+gz: <a href="https://genome.ucsc.edu/FAQ/FAQformat#format1">format</a>, 4th column required as gene identity.</li>
</ul>
</li>
<li><code>peakformat</code>: The format for <code>peakfile</code>, default: <code>peak</code>. It could be:  <ul>
<li>peak or peak+gz: (either <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format12">narrowPeak</a> or <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format13">broadPeak</a>, the 7th column will be used as intensity</li>
<li>bed or bed+gz: <a href="https://genome.ucsc.edu/FAQ/FAQformat#format1">format</a>, 5th column will be used as intensity.</li>
</ul>
</li>
<li><code>window</code>: <code>2 * d0</code> in the formula. The window where the peaks fall in will be consided, default: <code>100000</code>. <br />
<code>|--------- window ----------|
    |---- d0 -----|
    |--- 50K --- TSS --- 50K ---|
         ^ (peak center)
         |-- di --|</code></li>
</ul>
</li>
</ul>
</div>
<h3 id="cluster">cluster</h3>
<div class="admonition hint">
<p class="admonition-title">pDist2Coords</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert a distance matrix to coordinates, using multidimensional scaling.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The distance matrix, could be a full distance matrix, a triangle matrix or a pair-wise distance file  <ul>
<li>full dist matrix (full):
<div class="highlight"><pre><span></span>	s1	s2	s3
s1	0	1	1
s2	1	0	1
s3	1	1	0
</pre></div></li>
<li>triangle matrix (upper/lower), could be also lower triangle
<div class="highlight"><pre><span></span>	s1	s2	s3
s1	0	1	1
s2		0	1
s3			0
</pre></div></li>
<li>pair-wise (pair): (assuming auto-pair-wise distance = 0, that is: <code>s1 s1  0</code>)
<div class="highlight"><pre><span></span>s1	s2	1
s1	s3	1
s2	s3	1
</pre></div></li>
<li>Both rownames and header are required.</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output coordinate file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>informat</code>: The format of the input file: full, triangle or pair. Default: full  <ul>
<li>Could also be upper, lower, pair</li>
</ul>
</li>
<li><code>k</code>: How many dimensions? Default: 2 (R^2)  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCluster</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use <code>optCluster</code> to select the best number of clusters and cluster method, then perform the clustering</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input matrix file. Clustering will be performed against rows. If not, set <code>args.transpose</code> = True  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output cluster file  </li>
<li><code>outdir:dir</code>: The output directory containing the figures  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>transpose</code>: Transpose the input matrix. Default: False  </li>
<li><code>cnames</code>: Whether the input matrix contains header before transposing. Default: False  </li>
<li><code>rnames</code>: Which column is the rownames before transposing. Default: 1  </li>
<li><code>plot</code>: Whether plot the cluster. Default: True  </li>
<li><code>minc</code>: Min number of clusters to test. Default: 2  </li>
<li><code>maxc</code>: Min number of clusters to test. Default: 15  <ul>
<li>If number of rows (nrows) &lt;= 15, then max = nrows - 1</li>
</ul>
</li>
<li><code>methods</code>: The methods to test. Default: "all"  <ul>
<li>Could be any of "agnes", "clara", "diana", "fanny", "hierarchical", "kmeans", "model", "pam", "som", "sota", "em.nbinom", "da.nbinom", "sa.nbinom", "em.poisson", "da.poisson", "sa.poisson"</li>
<li>Multiple methods could be separated by comma (,), or put in a list</li>
<li>By default, fanny, model and sota will be excluded because fanny causes error and the latter two are slow. You can manually include them if you want.</li>
<li>Improper methods will be automatically excluded by <code>args.isCount</code></li>
</ul>
</li>
<li><code>isCount</code>: Whether the data is count data. Corresponding methods will be tested. Default: False  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://rdrr.io/cran/optCluster/man/optCluster.html"><code>r-optCluster</code></a>
    <a href="https://cran.r-project.org/web/packages/factoextra/index.html"><code>r-factoextra</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMCluster</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use <code>r-mclust</code> to do clustering. Current just do simple clustering with the package</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input a coordinate file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output of final results  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>transpose</code>: Transpose the input matrix? Default: False  </li>
<li><code>rnames</code>: The <code>row.names</code> for <code>read.table</code> to read the input file, default: True.  </li>
<li><code>cnames</code>: The <code>header</code> argument for <code>read.table</code> to read the input file, default: True.  </li>
<li><code>caption</code>: The caption for the <code>fviz_cluster</code>, default: "CLARA Clustering".  </li>
<li><code>minc</code>: The min # clusters to try, default: 2  </li>
<li><code>maxc</code>: The max # clusters to try, default: 15  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://cran.r-project.org/web/packages/mclust/index.html"><code>r-mclust</code></a>
    <a href="https://cran.r-project.org/web/packages/factoextra/index.html"><code>r-factoextra</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pAPCluster</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use <code>r-apcluster</code> to do clustering. </p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input a coordinate file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output of final results  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>transpose</code>: Transpose the input matrix? Default: False  </li>
<li><code>rownames</code>: The <code>row.names</code> for <code>read.table</code> to read the input file, default: 1.  </li>
<li><code>header</code>: The <code>header</code> argument for <code>read.table</code> to read the input file, default: True.  </li>
<li><code>caption</code>: The caption for the <code>fviz_cluster</code>, default: "APClustering".  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://cran.r-project.org/web/packages/apcluster/index.html"><code>r-apcluster</code></a>
    <a href="https://cran.r-project.org/web/packages/factoextra/index.html"><code>r-factoextra</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pHCluster</p>
<ul>
<li>
<p><strong>description</strong><br />
    Do hierarchical clustering.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input files with variants as rows, features as columns.  <ul>
<li>NOTE: clustering is performed on rows, rownames are the leaf labels.</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The result directory, containing:  <ul>
<li><code>hclust.merge.txt</code>: including merge and height information</li>
<li><code>hclust.order.txt</code>: including order and labels information</li>
<li><code>hclust.png</code>:       the dendrogram plot</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>fast</code>: whether to use <code>fastcluster</code> package or not, default: False  </li>
<li><code>gg</code>: whether to use <code>ggdendro</code> or not, default: False  </li>
<li><code>rownames</code>: The <code>row.names</code> for <code>read.table</code> to read the input file, default: 1.  </li>
<li><code>header</code>: The <code>header</code> argument for <code>read.table</code> to read the input file, default: True.  </li>
<li><code>method</code>: Which method to use for <code>hclust</code>. Default: "complete" (use <code>?hclust</code> to check all availables)  </li>
<li><code>rotate</code>: Which to rotate the plot or not. Default: False  </li>
<li><code>transpose</code>: Whether to transpose the matrix before cluster. Default: False  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://cran.r-project.org/web/packages/fastcluster/index.html"><code>r-fastcluster</code></a> if <code>args.fast</code> is True
    <a href="https://cran.r-project.org/web/packages/ggdendro/index.html"><code>r-ggdendro</code></a> if <code>args.gg</code> is True</p>
</li>
</ul>
</div>
<h3 id="cnvkit">cnvkit</h3>
<div class="admonition hint">
<p class="admonition-title">pCNVkitAccess</p>
<ul>
<li>
<p><strong>description</strong><br />
    Calculate the sequence-accessible coordinates in chromosomes from the given reference genome, output as a BED file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>fafile:file</code>: The fasta file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>params</code>: Other parameters for <code>cnvkit.py access</code>  </li>
<li><code>cnvkit</code>: The executable of cnvkit. Default: 'cnvkit.py'  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cnvkit.readthedocs.io/">CNVkit</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCNVkitTarget</p>
<ul>
<li>
<p><strong>description</strong><br />
    Generate targets file for CNVkit using access file and annotate file (<code>cnvkit.py target</code>)</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>acfile:file</code>: The access file  </li>
<li><code>anfile:file</code>: The annotate file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The targets file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnvkit</code>: The executable of cnvkit. Default: 'cnvkit.py'  </li>
<li><code>params</code>: Other parameters for <code>cnvkit.py target</code>  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cnvkit.readthedocs.io/">CNVkit</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCNVkitCov</p>
<ul>
<li>
<p><strong>description</strong><br />
    Calculate coverage in the given regions from BAM read depths.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The bam file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output cnn file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tgfile</code>: The target file  </li>
<li><code>cnvkit</code>: The executable of cnvkit. Default: 'cnvkit.py'  </li>
<li><code>nthread</code>: The number of threads to use. Default: 1  </li>
<li><code>params</code>: Other parameters for <code>cnvkit.py coverage</code>  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cnvkit.readthedocs.io/">CNVkit</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCNVkitRef</p>
<ul>
<li>
<p><strong>description</strong><br />
    Compile a copy-number reference from the given files or directory (containing normal samples). If given a reference genome (-f option), also calculate the GC content and repeat-masked proportion of each region.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: The input directory containing the cnn files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output reference cnn file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnvkit</code>: The executable of cnvkit. Default: 'cnvkit.py'  </li>
<li><code>params</code>: Other parameters for <code>cnvkit.py reference</code>, default: " --no-edge "  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cnvkit.readthedocs.io/">CNVkit</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCNVkitFix</p>
<ul>
<li>
<p><strong>description</strong><br />
    Combine the uncorrected target and antitarget coverage tables (.cnn) and correct for biases in regional coverage and GC content, according to the given reference. Output a table of copy number ratios (.cnr)</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The cnn file to be fixed  </li>
<li><code>rcfile:file</code>: The reference cnn file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The cnr file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnvkit</code>: The executable of cnvkit. Default: 'cnvkit.py'  </li>
<li><code>params</code>: Other parameters for <code>cnvkit.py fix</code>, default: " --no-edge "  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cnvkit.readthedocs.io/">CNVkit</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCNVkitSeg</p>
<ul>
<li>
<p><strong>description</strong><br />
    Infer discrete copy number segments from the given coverage table</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The cnr file   </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The cns file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnvkit</code>: The executable of cnvkit. Default: 'cnvkit.py'  </li>
<li><code>nthread</code>: The number of threads to use. Default: 1  </li>
<li><code>params</code>: Other parameters for <code>cnvkit.py segment</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cnvkit.readthedocs.io/">CNVkit</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCNVkitCall</p>
<ul>
<li>
<p><strong>description</strong><br />
    Given segmented log2 ratio estimates (.cns), derive each segment's absolute integer copy number </p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The cns file   </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The callcns file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnvkit</code>: The executable of cnvkit. Default: 'cnvkit.py'  </li>
<li><code>params</code>: Other parameters for <code>cnvkit.py segment</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cnvkit.readthedocs.io/">CNVkit</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCNVkitPlot</p>
<ul>
<li>
<p><strong>description</strong><br />
    Plot CNVkit results</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>cnrdir:file</code>: The directory containing copy number ratio files  </li>
<li><code>cnsdir:file</code>: The directory containing copy number segment files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnvkit</code>: The executable of cnvkit. Default: 'cnvkit.py'  </li>
<li><code>region</code>: The region for zoom-in plots. Default: '' (don't plot zoom-in view)  </li>
<li><code>gene</code>: The genes to be highlighted. Default: ''  </li>
<li><code>scatter</code>: Whether to generate the scatter plot. Default: True  </li>
<li><code>diagram</code>: Whether to generate the diagram plot. Default: True  </li>
<li><code>heatmap</code>: Whether to generate the heatmap plot. Default: True  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cnvkit.readthedocs.io/">CNVkit</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCNVkitRpt</p>
<ul>
<li>
<p><strong>description</strong><br />
    Report CNVkit results</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>cnrfile:file</code>: The file containing copy number ratio  </li>
<li><code>cnsfile:file</code>: The file containing copy number segment  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnvkit</code>: The executable of cnvkit. Default: 'cnvkit.py'  </li>
<li><code>breaks</code>: Whether to report breakpoints. Default: True  </li>
<li><code>gainloss</code>: Whether to report gainloss. Default: True  </li>
<li><code>metrics</code>: Whether to report metrics. Default: True  </li>
<li><code>segmetrics</code>: Whether to report segmetrics. Default: True  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cnvkit.readthedocs.io/">CNVkit</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCNVkit2Vcf</p>
<ul>
<li>
<p><strong>description</strong><br />
    Output vcf file for cnvkit results</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>cnsfile:file</code>: The cns file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The vcf file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnvkit</code>: The executable of cnvkit. Default: 'cnvkit.py'  </li>
<li><code>params</code>: Other params for <code>cnvkit.py export</code>  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cnvkit.readthedocs.io/">CNVkit</a></p>
</li>
</ul>
</div>
<h3 id="common">common</h3>
<div class="admonition hint">
<p class="admonition-title">pSort</p>
<ul>
<li>
<p><strong>description</strong><br />
    Sort file using linux command <code>sort</code></p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>skip</code>: To skip first N lines. Default: 0  </li>
<li><code>case</code>: Case-sensitivity. Default: True  <ul>
<li>If True, will set $LANG as C</li>
<li>Otherwise, $LANG will be set as en_US.UTF-8</li>
</ul>
</li>
<li><code>mem</code>    : The buffer size. Default: 4G  </li>
<li><code>tmpdir</code> : The tmpdir.  </li>
<li><code>unique</code> : Just keep the unique lines. Default: False  </li>
<li><code>delimit</code>: The delimit to separate the fields. Default: '\t'  </li>
<li><code>params</code> : The arguments used by <code>sort</code>  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pFiles2Dir</p>
<ul>
<li>
<p><strong>description</strong><br />
    A helper process to convert a list of files into a directory, so that some processes can take it as input</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infiles:files</code>: The input files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pFile2Proc</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert a file to a proc so it can be used as dependent</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pStr2File</p>
<ul>
<li>
<p><strong>description</strong><br />
    Save string to a file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>in:var</code>: The input string.  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file.  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pHead</p>
<ul>
<li>
<p><strong>description</strong><br />
    Get the top N lines from a file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>n</code>: Top n lines. You may use '-n' to skip last n lines.  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pTail</p>
<ul>
<li>
<p><strong>description</strong><br />
    Get the bottom N lines from a file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>n</code>: Bottom n lines. You may use '+n' to skip first n lines.  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pPrepend</p>
<ul>
<li>
<p><strong>description</strong><br />
    Prepend a string to a file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>in:var</code>: The input string.  </li>
<li><code>infile:file</code>: The input file.  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file.  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pAddHeader</p>
<ul>
<li>
<p><strong>description</strong><br />
    Add the header of 1st file to 2nd file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile1:file</code>: The first file containing the header.  </li>
<li><code>infile2:file</code>: The second file with the body.  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file with the header from 1st input file, body from 2nd file.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>n</code>: The number of header lines.  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMergeFiles</p>
<ul>
<li>
<p><strong>description</strong><br />
    Merge files in the input directory</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: The input directory  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>inopts</code>: The options for input file.  <ul>
<li>defaults: skip: 0, comment: #, delimit '\t'</li>
</ul>
</li>
<li><code>outopts</code>: The options for output file. Defaults:  <ul>
<li>head: False (not output head line)</li>
<li>headPrefix: <code>#</code> (The prefix for head line)</li>
<li>headDelimit: <code>\\t</code> (The delimit for head line)</li>
<li>headTransform: <code>None</code> (The callback for head line)</li>
<li>delimit: <code>\\t</code> (The delimit for output line)</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSplitRows</p>
<ul>
<li>
<p><strong>description</strong><br />
    Split a file by rows, specially usefull to split a job into multithreads/multiprocesses.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory including the split files  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>skip</code>: The skip first n lines. Default: <code>0</code>  </li>
<li><code>cnames</code>: The column names. If True, the column names will be added to each split file. Default: <code>True</code>  </li>
<li><code>n</code>: Number of files to split. Default: <code>8</code>  </li>
</ul>
</li>
</ul>
</div>
<h3 id="eqtl">eqtl</h3>
<div class="admonition hint">
<p class="admonition-title">pMatrixeQTL</p>
<ul>
<li>
<p><strong>description</strong><br />
    Call eQTLs using Matrix eQTL</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>snpfile:file</code>: The genotype file, rows are snps and columns are samples  </li>
<li><code>expfile:file</code>: The expression file, rows are genes  </li>
<li><code>covfile:file</code>: The covariant file, rows are covariants  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The matrix eqtl output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>model</code>: The model to use, either modelLINEAR(default) or modelANOVA  </li>
<li><code>pval</code> : The pvalue cutoff (if <code>cisopts.dist</code> &gt; 0, will be used as pval for trans-eQTL)  </li>
<li><code>fdr</code>  : Calculate FDR or not (default: True)  </li>
<li><code>cisopts</code>: Options for calling cis-, trans-eQTL  <ul>
<li><code>snppos</code> : The snp position file (columns are: snp, chr, pos)</li>
<li><code>genepos</code>: The gene position file (columns are: gene, chr, start, end)</li>
<li><code>dist</code>   : The distance to define cis-eQTL. (default: 0 (don't do cis-, trans- calling)</li>
<li><code>cispv</code>  : The pvalue cutoff for cis-eQTL (<code>pval</code> will not work)</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://www.bios.unc.edu/research/genomic_software/Matrix_eQTL/"><code>Matrix-eQTL (R)</code></a>     </p>
</li>
</ul>
</div>
<h3 id="fastx">fastx</h3>
<div class="admonition hint">
<p class="admonition-title">pFastq2Expr</p>
<ul>
<li><strong>description</strong><br />
    Use Kallisto to get gene expression from pair-end fastq files.</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pFastqSim</p>
<ul>
<li>
<p><strong>description</strong><br />
    Simulate reads</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>seed</code>: The seed to generate simulation file  <ul>
<li>None: use current timestamp.</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>fq1:file</code>: The first pair read file  </li>
<li><code>fq2:file</code>: The second pair read file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used for simulation. Default: wgsim (dwgsim)  </li>
<li><code>len1</code>: The length of first pair read. Default: 100  </li>
<li><code>len2</code>: The length of second pair read. Default: 100  </li>
<li><code>num</code>: The number of read PAIRs. Default: 1000000  </li>
<li><code>gz</code>: Whether generate gzipped read file. Default: True  </li>
<li><code>wgsim</code>: The path of wgsim. Default: wgsim  </li>
<li><code>dwgsim</code>: The path of wgsim. Default: dwgsim  </li>
<li><code>ref</code>: The reference genome. Required  </li>
<li><code>params</code>: Other params for <code>tool</code>. Default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/lh3/wgsim"><code>wgsim</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pFastQC</p>
<ul>
<li>
<p><strong>description</strong><br />
    QC report for fastq file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>fq:file</code>: The fastq file (also fine with gzipped)  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output direcotry  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used for simulation. Default: fastqc  </li>
<li><code>fastqc</code>: The path of fastqc. Default: fastqc  </li>
<li><code>nthread</code>: Number of threads to use. Default: 1  </li>
<li><code>params</code>: Other params for <code>tool</code>. Default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/"><code>fastqc</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pFastMC</p>
<ul>
<li>
<p><strong>description</strong><br />
    Multi-QC based on pFastQC</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>qcdir:file</code>: The direcotry containing QC files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output direcotry  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used for simulation. Default: multiqc  </li>
<li><code>multiqc</code>: The path of fastqc. Default: multiqc  </li>
<li><code>params</code>: Other params for <code>tool</code>. Default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://multiqc.info/"><code>multiqc</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pFastqTrim</p>
<ul>
<li>
<p><strong>description</strong><br />
    Trim pair-end FASTQ reads</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>fq1:file</code>: The input fastq file  </li>
<li><code>fq2:file</code>: The input fastq file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfq1:file</code>: The trimmed fastq file  </li>
<li><code>outfq2:file</code>: The trimmed fastq file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>        : The tools used for trimming. Default: trimmomatic (cutadapt|skewer)  </li>
<li><code>cutadapt</code>    : The path of seqtk. Default: cutadapt  </li>
<li><code>skewer</code>      : The path of fastx toolkit trimmer. Default: skewer  </li>
<li><code>trimmomatic</code> : The path of trimmomatic. Default: trimmomatic  </li>
<li><code>params</code>      : Other params for <code>tool</code>. Default: ""  </li>
<li><code>nthread</code>     : Number of threads to be used. Default: 1  </li>
<li>Not for cutadapt</li>
<li><code>gz</code>          : Whether gzip output files. Default: True  </li>
<li><code>mem</code>         : The memory to be used. Default: 4G  </li>
<li>Only for trimmomatic</li>
<li><code>minlen</code>      : Discard trimmed reads that are shorter than <code>minlen</code>. Default: 18  </li>
<li>For trimmomatic, the number will be <code>minlen</code>*2 for MINLEN, as it filters before trimming</li>
<li><code>minq</code>        : Minimal mean qulity for 4-base window or leading/tailing reads. Default: 3  </li>
<li><code>cut5</code>        : Remove the 5'end reads if they are below qulity. Default: 3  </li>
<li><code>cut3</code>        : Remove the 3'end reads if they are below qulity. Default: 3  </li>
<li>Not for skewer</li>
<li><code>adapter1</code>    : The adapter for sequence. Default: AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC  </li>
<li><code>adapter2</code>    : The adapter for pair-end sequence. Default: AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cutadapt.readthedocs.io/en/stable/guide.html"><code>cutadapt</code></a>
    <a href="https://github.com/relipmoc/skewer"><code>skewer</code></a>
    <a href="https://github.com/timflutre/trimmomatic"><code>trimmomatic</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pFastqSETrim</p>
<ul>
<li>
<p><strong>description</strong><br />
    Trim single-end FASTQ reads</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>fq:file</code>: The input fastq file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfq:file</code>: The trimmed fastq file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>        : The tools used for trimming. Default: trimmomatic (cutadapt|skewer)  </li>
<li><code>cutadapt</code>    : The path of seqtk. Default: cutadapt  </li>
<li><code>skewer</code>      : The path of fastx toolkit trimmer. Default: skewer  </li>
<li><code>trimmomatic</code> : The path of trimmomatic. Default: trimmomatic  </li>
<li><code>params</code>      : Other params for <code>tool</code>. Default: ""  </li>
<li><code>nthread</code>     : Number of threads to be used. Default: 1  </li>
<li>Not for cutadapt</li>
<li><code>gz</code>          : Whether gzip output files. Default: True  </li>
<li><code>mem</code>         : The memory to be used. Default: 4G  </li>
<li>Only for trimmomatic</li>
<li><code>minlen</code>      : Discard trimmed reads that are shorter than <code>minlen</code>. Default: 18  </li>
<li>For trimmomatic, the number will be <code>minlen</code>*2 for MINLEN, as it filters before trimming</li>
<li><code>minq</code>        : Minimal mean qulity for 4-base window or leading/tailing reads. Default: 3  </li>
<li><code>cut5</code>        : Remove the 5'end reads if they are below qulity. Default: 3  </li>
<li><code>cut3</code>        : Remove the 3'end reads if they are below qulity. Default: 3  </li>
<li>Not for skewer</li>
<li><code>adapter</code>     : The adapter for sequence. Default: AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cutadapt.readthedocs.io/en/stable/guide.html"><code>cutadapt</code></a>
    <a href="https://github.com/relipmoc/skewer"><code>skewer</code></a>
    <a href="https://github.com/timflutre/trimmomatic"><code>trimmomatic</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pFastqSE2Sam</p>
<ul>
<li>
<p><strong>description</strong><br />
    Cleaned paired fastq (.fq, .fq.gz, .fastq, .fastq.gz file to mapped sam/bam file</p>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used for alignment. Default: bwa (bowtie2|ngm)  </li>
<li><code>bwa</code>: Path of bwa, default: bwa  </li>
<li><code>ngm</code>: Path of ngm, default: ngm  </li>
<li><code>bowtie2</code>: Path of bowtie2, default: bowtie2  </li>
<li><code>rg</code>: The read group. Default: {'id': '', 'pl': 'Illumina', 'pu': 'unit1', 'lb': 'lib1', 'sm': ''}  </li>
<li><code>id</code> will be parsed from filename with "<em>LX</em>" in it if not given</li>
<li><code>sm</code> will be parsed from filename</li>
<li><code>ref</code>: Path of reference file  </li>
<li><code>params</code>: Other params for tool, default: ''  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pFastq2Sam</p>
<ul>
<li>
<p><strong>description</strong><br />
    Cleaned paired fastq (.fq, .fq.gz, .fastq, .fastq.gz file to mapped sam/bam file</p>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>   : The tool used for alignment. Default: bwa (bowtie2, ngm, star)  </li>
<li><code>bwa</code>    : Path of bwa, default: bwa  </li>
<li><code>ngm</code>    : Path of ngm, default: ngm  </li>
<li><code>star</code>   : Path of ngm, default: STAR  </li>
<li><code>bowtie2</code>: Path of bowtie2, default: bowtie2  </li>
<li><code>rg</code>: The read group. Default: {'id': '', 'pl': 'Illumina', 'pu': 'unit1', 'lb': 'lib1', 'sm': ''}  </li>
<li><code>id</code> will be parsed from filename with "<em>LX</em>" in it if not given</li>
<li><code>sm</code> will be parsed from filename</li>
<li><code>ref</code>    : Path of reference file  </li>
<li><code>refgene</code>: The GTF file for STAR to build index. It's not neccessary if index is already been built. Default: ''  </li>
<li><code>params</code> : Other params for tool, default: ''  </li>
</ul>
</li>
</ul>
</div>
<h3 id="gatk">gatk</h3>
<div class="admonition hint">
<p class="admonition-title">pRealignerTargetCreator</p>
<ul>
<li>
<p><strong>description</strong><br />
    The local realignment process is designed to consume one or more BAM files and to locally realign reads such that the number of mismatching bases is minimized across all the reads. In general, a large percent of regions requiring local realignment are due to the presence of an insertion or deletion (indels) in the individual's genome with respect to the reference genome. Such alignment artifacts result in many bases mismatching the reference near the misalignment, which are easily mistaken as SNPs. Moreover, since read mapping algorithms operate on each read independently, it is impossible to place reads on the reference genome such that mismatches are minimized across all reads. Consequently, even when some reads are correctly mapped with indels, reads covering the indel near just the start or end of the read are often incorrectly mapped with respect the true indel, also requiring realignment. Local realignment serves to transform regions with misalignments due to indels into clean reads containing a consensus indel suitable for standard variant discovery approaches.
    Note that indel realignment is no longer necessary for variant discovery if you plan to use a variant caller that performs a haplotype assembly step, such as HaplotypeCaller or MuTect2. However it is still required when using legacy callers such as UnifiedGenotyper or the original MuTect. There are 2 steps to the realignment process:</p>
<ul>
<li>Determining (small) suspicious intervals which are likely in need of realignment (RealignerTargetCreator)</li>
<li>Running the realigner over those intervals (see the IndelRealigner tool)
For more details, see <a href="http://www.broadinstitute.org/gatk/guide/article?id=38">the indel realignment method documentation</a>.</li>
</ul>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>bamfile:file</code>: The aligned bam file  </li>
<li><code>reffile</code>: The reference file  </li>
</ul>
</li>
<li>
<p><strong>brings</strong>  </p>
<ul>
<li><code>bamfile</code>: <code>{{bamfile | bn}}.bai</code> The index file of input bam file  </li>
<li><code>reffile#fai</code>: <code>{{reffile | bn}}.fai</code>  </li>
<li><code>reffile#dict</code>: <code>{{reffile | bn}}.dict</code>  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: A list of target intervals to pass to the IndelRealigner.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>gatk</code>: The gatk executable, default: "gatk"  </li>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>params</code>: Other parameters for RealignerTargetCreator, default: ""  </li>
<li><code>samtools</code>: The samtools executable, default: "samtools"  </li>
<li><code>tmpdir</code>: The tmpdir to use. Default: /tmp  </li>
<li><code>javamem</code>: The memory for java vm. Default: "-Xms1g -Xmx8g"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://software.broadinstitute.org/gatk">GATK</a>
    <a href="http://www.htslib.org/">samtools</a> if <code>reffile</code> is not indexed or <code>bamfile</code> is not indexed.
    <a href="https://broadinstitute.github.io/picard/">picard</a> if <code>reffile</code> is not dicted.</p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pIndelRealigner</p>
<ul>
<li>
<p><strong>description</strong><br />
    The local realignment process is designed to consume one or more BAM files and to locally realign reads such that the number of mismatching bases is minimized across all the reads. In general, a large percent of regions requiring local realignment are due to the presence of an insertion or deletion (indels) in the individual's genome with respect to the reference genome. Such alignment artifacts result in many bases mismatching the reference near the misalignment, which are easily mistaken as SNPs. Moreover, since read mapping algorithms operate on each read independently, it is impossible to place reads on the reference genome such at mismatches are minimized across all reads. Consequently, even when some reads are correctly mapped with indels, reads covering the indel near just the start or end of the read are often incorrectly mapped with respect the true indel, also requiring realignment. Local realignment serves to transform regions with misalignments due to indels into clean reads containing a consensus indel suitable for standard variant discovery approaches.
    Note that indel realignment is no longer necessary for variant discovery if you plan to use a variant caller that performs a haplotype assembly step, such as HaplotypeCaller or MuTect2. However it is still required when using legacy callers such as UnifiedGenotyper or the original MuTect.
    There are 2 steps to the realignment process:</p>
<ul>
<li>Determining (small) suspicious intervals which are likely in need of realignment (see the RealignerTargetCreator tool)</li>
<li>Running the realigner over those intervals (IndelRealigner)
For more details, see <a href="http://www.broadinstitute.org/gatk/guide/article?id=38">the indel realignment method documentation</a>.</li>
</ul>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>bamfile:file</code>: The aligned bam file  </li>
<li><code>intfile:file</code>: Intervals file output from RealignerTargetCreator  </li>
<li><code>reffile:file</code>: The reference file  </li>
</ul>
</li>
<li>
<p><strong>brings</strong>  </p>
<ul>
<li><code>bamfile</code>: <code>{{bamfile | bn}}.bai</code> The index file of input bam file  </li>
<li><code>reffile#fai</code>: <code>{{reffile | bn}}.fai</code>  </li>
<li><code>reffile#dict</code>: <code>{{reffile | bn}}.dict</code>  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: A realigned version of input BAM file.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>gatk</code>: The gatk executable, default: "gatk"  </li>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>params</code>: Other parameters for IndelRealigner, default: ""  </li>
<li><code>samtools</code>: The samtools executable, default: samtools  </li>
<li><code>tmpdir</code>: The tmpdir to use. Default: /tmp  </li>
<li><code>javamem</code>: The memory for java vm. Default: "-Xms1g -Xmx8g"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://software.broadinstitute.org/gatk">GATK</a>
    <a href="http://www.htslib.org/">samtools</a> if <code>reffile</code> is not indexed or <code>bamfile</code> is not indexed.
    <a href="https://broadinstitute.github.io/picard/">picard</a> if <code>reffile</code> is not dicted.</p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBaseRecalibrator</p>
<ul>
<li>
<p><strong>description</strong><br />
    Variant calling algorithms rely heavily on the quality scores assigned to the individual base calls in each sequence read. These scores are per-base estimates of error emitted by the sequencing machines. Unfortunately the scores produced by the machines are subject to various sources of systematic technical error, leading to over- or under-estimated base quality scores in the data. Base quality score recalibration (BQSR) is a process in which we apply machine learning to model these errors empirically and adjust the quality scores accordingly. This allows us to get more accurate base qualities, which in turn improves the accuracy of our variant calls. The base recalibration process involves two key steps: first the program builds a model of covariation based on the data and a set of known variants (which you can bootstrap if there is none available for your organism), then it adjusts the base quality scores in the data based on the model. There is an optional but highly recommended step that involves building a second model and generating before/after plots to visualize the effects of the recalibration process. This is useful for quality control purposes. This tool performs the first step described above: it builds the model of covariation and produces the recalibration table. It operates only at sites that are not in dbSNP; we assume that all reference mismatches we see are therefore errors and indicative of poor base quality. This tool generates tables based on various user-specified covariates (such as read group, reported quality score, cycle, and context). Assuming we are working with a large amount of data, we can then calculate an empirical probability of error given the particular covariates seen at this site, where p(error) = num mismatches / num observations. The output file is a table (of the several covariate values, number of observations, number of mismatches, empirical quality score).</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>bamfile:file</code>: A BAM file containing data that needs to be recalibrated.  </li>
<li><code>reffile:file</code>: The reference file  </li>
</ul>
</li>
<li>
<p><strong>brings</strong>  </p>
<ul>
<li><code>bamfile</code>: <code>{{bamfile | bn}}.bai</code> The index file of input bam file  </li>
<li><code>reffile#fai</code>: <code>{{reffile | bn}}.fai</code>  </li>
<li><code>reffile#dict</code>: <code>{{reffile | bn}}.dict</code>  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: A GATKReport file with many tables:  <ul>
<li>The list of arguments</li>
<li>The quantized qualities table</li>
<li>The recalibration table by read group</li>
<li>The recalibration table by quality score</li>
<li>The recalibration table for all the optional covariates</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>gatk</code>: The gatk executable, default: "gatk"  </li>
<li><code>params</code>: Other parameters for BaseRecalibrator, default: ""  </li>
<li><code>knownSites</code>: The known polymorphic sites to mask out, required  </li>
<li><code>samtools</code>: The samtools executable, default: samtools  </li>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>tmpdir</code>: The tmpdir to use. Default: /tmp  </li>
<li><code>javamem</code>: The memory for java vm. Default: "-Xms1g -Xmx8g"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://software.broadinstitute.org/gatk">GATK</a>
    <a href="http://www.htslib.org/">samtools</a> if <code>reffile</code> is not indexed or <code>bamfile</code> is not indexed.
    <a href="https://broadinstitute.github.io/picard/">picard</a> if <code>reffile</code> is not dicted.</p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pPrintReads</p>
<ul>
<li>
<p><strong>description</strong><br />
    PrintReads is a generic utility tool for manipulating sequencing data in SAM/BAM format. It can dynamically merge the contents of multiple input BAM files, resulting in merged output sorted in coordinate order. It can also optionally filter reads based on various read properties such as read group tags using the <code>--read_filter/-rf</code> command line argument (see documentation on read filters for more information).
    Note that when PrintReads is used as part of the Base Quality Score Recalibration workflow, it takes the <code>--BQSR</code> engine argument, which is listed under Inherited Arguments &gt; CommandLineGATK below.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>bamfile:file</code>: A BAM file.  </li>
<li><code>recaltable:file</code>: The GATKReport file  </li>
<li><code>reffile:file</code>: The reference file  </li>
</ul>
</li>
<li>
<p><strong>brings</strong>  </p>
<ul>
<li><code>bamfile</code>: <code>{{bamfile | bn}}.bai</code> The index file of input bam file  </li>
<li><code>reffile#fai</code>: <code>{{reffile | bn}}.fai</code>  </li>
<li><code>reffile#dict</code>: <code>{{reffile | bn}}.dict</code>  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: A single processed bam file.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>gatk</code>: The gatk executable, default: "gatk"  </li>
<li><code>params</code>: Other parameters for PrintReads, default: ""  </li>
<li><code>samtools</code>: The samtools executable, default: samtools  </li>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>tmpdir</code>: The tmpdir to use. Default: /tmp  </li>
<li><code>javamem</code>: The memory for java vm. Default: "-Xms1g -Xmx8g"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://software.broadinstitute.org/gatk">GATK</a>
    <a href="http://www.htslib.org/">samtools</a> if <code>reffile</code> is not indexed or <code>infile</code> is not indexed.
    <a href="https://broadinstitute.github.io/picard/">picard</a> if <code>reffile</code> is not dicted.</p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pHaplotypeCaller</p>
<ul>
<li>
<p><strong>description</strong><br />
    PrintReads is a generic utility tool for manipulating sequencing data in SAM/BAM format. It can dynamically merge the contents of multiple input BAM files, resulting in merged output sorted in coordinate order. It can also optionally filter reads based on various read properties such as read group tags using the <code>--read_filter/-rf</code> command line argument (see documentation on read filters for more information).
    Note that when PrintReads is used as part of the Base Quality Score Recalibration workflow, it takes the <code>--BQSR</code> engine argument, which is listed under Inherited Arguments &gt; CommandLineGATK below.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>bamfile:file</code>: A BAM file.  </li>
<li><code>reffile:file</code>: The reference file  </li>
</ul>
</li>
<li>
<p><strong>brings</strong>  </p>
<ul>
<li><code>bamfile</code>: <code>{{bamfile | bn}}.bai</code> The index file of input bam file  </li>
<li><code>reffile#fai</code>: <code>{{reffile | bn}}.fai</code>  </li>
<li><code>reffile#dict</code>: <code>{{reffile | bn}}.dict</code>  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: Either a VCF or gVCF file with raw, unfiltered SNP and indel calls.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>gatk</code>    : The gatk executable, default: "gatk"  </li>
<li><code>params</code>  : Other parameters for HaplotypeCaller, default: ""  </li>
<li><code>samtools</code>: The samtools executable, default: samtools  </li>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>tmpdir</code>: The tmpdir to use. Default: /tmp  </li>
<li><code>javamem</code>: The memory for java vm. Default: "-Xms1g -Xmx8g"  </li>
<li><code>nthread</code>: Corresponding to -nct option  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://software.broadinstitute.org/gatk">GATK</a>
    <a href="http://www.htslib.org/">samtools</a> if <code>reffile</code> is not indexed or <code>infile</code> is not indexed.
    <a href="https://broadinstitute.github.io/picard/">picard</a> if <code>reffile</code> is not dicted.</p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSelectVariants</p>
<ul>
<li>
<p><strong>description</strong><br />
    Often, a VCF containing many samples and/or variants will need to be subset in order to facilitate certain analyses (e.g. comparing and contrasting cases vs. controls; extracting variant or non-variant loci that meet certain requirements, displaying just a few samples in a browser like IGV, etc.). SelectVariants can be used for this purpose.
    There are many different options for selecting subsets of variants from a larger callset:</p>
<ul>
<li>Extract one or more samples from a callset based on either a complete sample name or a pattern match.</li>
<li>Specify criteria for inclusion that place thresholds on annotation values, e.g. "DP &gt; 1000" (depth of coverage greater than 1000x), "AF &lt; 0.25" (sites with allele frequency less than 0.25). These - criteria are written as "JEXL expressions", which are documented in the article about using JEXL expressions.</li>
<li>Provide concordance or discordance tracks in order to include or exclude variants that are also present in other given callsets.</li>
<li>Select variants based on criteria like their type (e.g. INDELs only), evidence of mendelian violation, filtering status, allelicity, and so on.
There are also several options for recording the original values of certain annotations that are recalculated when a subsetting the new callset, trimming alleles, and so on.</li>
</ul>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>vcffile:file</code>: A variant call set from which to select a subset.  </li>
<li><code>reffile:file</code>: The reference file  </li>
</ul>
</li>
<li>
<p><strong>brings</strong>  </p>
<ul>
<li><code>reffile#fai</code>: <code>{{reffile | bn}}.fai</code>  </li>
<li><code>reffile#dict</code>: <code>{{reffile | bn}}.dict</code>  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: A new VCF file containing the selected subset of variants.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>gatk</code>: The gatk executable, default: "gatk"  </li>
<li><code>params</code>: Other parameters for SelectVariants, default: ""  </li>
<li><code>samtools</code>: The samtools executable, default: samtools  </li>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>tmpdir</code>: The tmpdir to use. Default: /tmp  </li>
<li><code>javamem</code>: The memory for java vm. Default: "-Xms1g -Xmx8g"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://software.broadinstitute.org/gatk">GATK</a>
    <a href="http://www.htslib.org/">samtools</a> if <code>reffile</code> is not indexed or <code>infile</code> is not indexed.
    <a href="https://broadinstitute.github.io/picard/">picard</a> if <code>reffile</code> is not dicted.</p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pVariantFiltration</p>
<ul>
<li>
<p><strong>description</strong><br />
    This tool is designed for hard-filtering variant calls based on certain criteria. Records are hard-filtered by changing the value in the FILTER field to something other than PASS. Filtered records will be preserved in the output unless their removal is requested in the command line.
    The most common way of specifying filtering criteria is by using JEXL queries. See the article on JEXL expressions in the documentation Guide for detailed information and examples.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>vcffile:file</code>: A variant call set from which to select a subset.  </li>
<li><code>reffile:file</code>: The reference file  </li>
</ul>
</li>
<li>
<p><strong>brings</strong>  </p>
<ul>
<li><code>reffile#fai</code>: <code>{{reffile | bn}}.fai</code>  </li>
<li><code>reffile#dict</code>: <code>{{reffile | bn}}.dict</code>  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: A filtered VCF.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>gatk</code>: The gatk executable, default: "gatk -T VariantFiltration"  </li>
<li><code>params</code>: Other parameters for VariantFiltration, default: ""  </li>
<li><code>samtools</code>: The samtools executable, default: samtools  </li>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>tmpdir</code>: The tmpdir to use. Default: /tmp  </li>
<li><code>javamem</code>: The memory for java vm. Default: "-Xms1g -Xmx8g"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://software.broadinstitute.org/gatk">GATK</a>
    <a href="http://www.htslib.org/">samtools</a> if <code>reffile</code> is not indexed or <code>infile</code> is not indexed.
    <a href="https://broadinstitute.github.io/picard/">picard</a> if <code>reffile</code> is not dicted.</p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMuTect2</p>
<ul>
<li>
<p><strong>description</strong><br />
    MuTect2 is a somatic SNP and indel caller that combines the DREAM challenge-winning somatic genotyping engine of the original MuTect (<a href="http://www.nature.com/nbt/journal/v31/n3/full/nbt.2514.html">Cibulskis et al., 2013</a>) with the assembly-based machinery of HaplotypeCaller. The basic operation of MuTect2 proceeds similarly to that of the HaplotypeCaller.
    NOTE: only Tumor/Normal variant calling implemented in bioprocs</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>tumor:file</code>: the tumor bam file  </li>
<li><code>normal:file</code>: the normal bam file  </li>
<li><code>reffile:file</code>: the reference file  </li>
</ul>
</li>
<li>
<p><strong>brings</strong>  </p>
<ul>
<li><code>tumor</code>: <code>{{tumor | bn}}.bai</code> the index file of tumor  </li>
<li><code>normal</code>: <code>{{normal | bn}}.bai</code> the index file of normal  </li>
<li><code>reffile#fai</code>: <code>{{reffile | bn}}.fai</code>  </li>
<li><code>reffile#dict</code>: <code>{{reffile | bn}}.dict</code>  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The vcf file containing somatic mutations  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>gatk</code>: The gatk executable, default: "gatk"  </li>
<li><code>samtools</code>: The samtools executable, default: samtools  </li>
<li><code>params</code>: Other parameters for MuTect2, default: ""  </li>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>tmpdir</code>: The tmpdir to use. Default: /tmp  </li>
<li><code>javamem</code>: The memory for java vm. Default: "-Xms1g -Xmx8g"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://software.broadinstitute.org/gatk">GATK</a>
    <a href="http://www.htslib.org/">samtools</a> if index files of input files are not found
    <a href="https://broadinstitute.github.io/picard/">picard</a> if <code>reffile</code> is not dicted.</p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMuTect2Interval</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use interval file model of MuTect2</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>tumor:file</code>: the tumor bam file  </li>
<li><code>normal:file</code>: the normal bam file  </li>
<li><code>reffile:file</code>: the reference file  </li>
</ul>
</li>
<li>
<p><strong>brings</strong>  </p>
<ul>
<li><code>tumor</code>: <code>{{tumor | bn}}.bai</code> the index file of tumor  </li>
<li><code>normal</code>: <code>{{normal | bn}}.bai</code> the index file of normal  </li>
<li><code>reffile#fai</code>: <code>{{reffile | bn}}.fai</code>  </li>
<li><code>reffile#dict</code>: <code>{{reffile | bn}}.dict</code>  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The vcf file containing somatic mutations  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>gatk</code>: The gatk executable, default: "gatk"  </li>
<li><code>samtools</code>: The samtools executable, default: samtools  </li>
<li><code>params</code>: Other parameters for MuTect2, default: ""  </li>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>tmpdir</code>: The tmpdir to use. Default: /tmp  </li>
<li><code>javamem</code>: The memory for java vm. Default: "-Xms1g -Xmx8g"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://software.broadinstitute.org/gatk">GATK</a>
    <a href="http://www.htslib.org/">samtools</a> if index files of input files are not found
    <a href="https://broadinstitute.github.io/picard/">picard</a> if <code>reffile</code> is not dicted.</p>
</li>
</ul>
</div>
<h3 id="gene">gene</h3>
<div class="admonition hint">
<p class="admonition-title">pGenePromoters</p>
<ul>
<li><strong>description</strong><br />
    Alias of <code>seq.pPromoters</code>.</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pGeneNameNorm</p>
<ul>
<li>
<p><strong>description</strong><br />
    Normalize gene names using MyGeneinfo.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>notfound</code>: What if a symbol is not found. Default: ignore  <ul>
<li>skip  : skip the record(don't write it to output file)</li>
<li>ignore: use the original name;</li>
<li>error : report error</li>
</ul>
</li>
<li><code>col</code>: the column index containing the gene names  </li>
<li><code>from</code>: the original format. Default: 'symbol, alias'  </li>
<li><code>to</code>: the output gene name format. Default: 'symbol'  </li>
<li><code>genome</code>: the genome. Default: 'hg19'  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pGeneTss</p>
<ul>
<li>
<p><strong>description</strong><br />
    Get gene TSS in BEd format.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file containing genes  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output BED file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>notfound</code>: What if the gene is not found. Default: skip.  <ul>
<li>error: report error</li>
</ul>
</li>
<li><code>header</code>: Whether the input file contains header. Default: False  </li>
<li><code>skip</code>: Skip N lines of input file. Default: 0  <ul>
<li>This has highest priority of header and comment</li>
</ul>
</li>
<li><code>comment</code>: The comment line start sign. Default: #  </li>
<li><code>delimit</code>: The delimit of input file if it has multiple column. Default: <code>\\t</code>  </li>
<li><code>col</code>: The column index contains the genes. Default: 0  </li>
<li><code>frm</code>: The format of the genes. Default: <code>symbol, alias</code>  </li>
<li><code>tmpdir</code>: The tmpdir used to store mygene cache files.  </li>
<li><code>genome</code>: In which genome to fetch the coordinates. Default: hg19  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pGeneBody</p>
<ul>
<li>
<p><strong>description</strong><br />
    Get gene body region in BED format</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file containing genes  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The gene body region  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>notfound</code>: What if a gene is not found when transfer the gene names to gene symbols  <ul>
<li>error: report error</li>
<li>skip (default): skip it</li>
</ul>
</li>
<li><code>inmeta</code>: The metadata for input file, mainly to indicate where the GENE column is.  </li>
<li><code>inopts</code>: Input options for reading input file.  <ul>
<li>skip: number of lines to skip. Default: 0</li>
<li>comment: the starting string for comment lines. Default: #</li>
<li>delimit: The delimit for the input file. Default: '\t'
frm: The gene name format in the input file. Default: 'symbol, alias'
tmpdir: The tmpdir to cache the gene name conversion.
genome: The genome used to do the conversion.</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<h3 id="genomeplot">genomeplot</h3>
<div class="admonition hint">
<p class="admonition-title">pInteractionTrack</p>
<ul>
<li>
<p><strong>description</strong><br />
    Gererate genomic interaction track for Gviz</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>name</code>: The name of the track  </li>
<li><code>infile:file</code>: The input file.   <ul>
<li>See the <code>type</code> argument for <code>makeGenomicInteractionsFromFile</code> from <code>GenomicInteractions</code> r-package</li>
</ul>
</li>
<li><code>region</code>: the region, just chromosome!  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The dumped track data  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>intype</code>: Input file type. Default: auto  <ul>
<li>Identified by extension</li>
<li>One of "chiapet.tool", "bed12", "bedpe", "hiclib", "homer", "bam", "two.bams".</li>
</ul>
</li>
<li><code>params</code>: The display params  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pGeneTrack</p>
<ul>
<li>
<p><strong>description</strong><br />
    Generate gene track using ucsc data source</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>name</code>: The name of the track  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The file to save the track  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>genome</code>: The genome  </li>
<li><code>params</code>: use <code>displayPars(UcscTrack(genome="mm9", chromosome="chrM", track="knownGene"))</code> to see all available args  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://rdrr.io/bioc/Gviz">r-Gviz</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pAnnoTrack</p>
<ul>
<li>
<p><strong>description</strong><br />
    The annotation track of Gviz</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>name</code>: the name of the track  </li>
<li><code>infile:file</code>: the file for the track. (wig, bigWig or bedGraph, bam, need to be indexed!)  </li>
<li><code>chrom</code>: the chrom  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the rds file for the track  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>genome</code>: The genome  </li>
<li><code>params</code>: See <code>displayPars(DataTrack())</code> for all available display params  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://rdrr.io/bioc/Gviz/man/DataTrack-class.html">r-Gviz</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pDataTrack</p>
<ul>
<li>
<p><strong>description</strong><br />
    The data track of Gviz</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>name</code>: the name of the track  </li>
<li><code>infile:file</code>: the file for the track. (wig, bigWig or bedGraph, bam, need to be indexed!)  </li>
<li><code>chrom</code>: the chrom  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the rds file for the track  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>genome</code>: The genome  </li>
<li><code>params</code>: See <code>displayPars(DataTrack())</code> for all available display params  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://rdrr.io/bioc/Gviz/man/DataTrack-class.html">r-Gviz</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pUcscTrack</p>
<ul>
<li>
<p><strong>description</strong><br />
    Generate track from ucsc</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>name</code>     : the name of the track  </li>
<li><code>track</code>    : the UCSC track  </li>
<li><code>trackType</code>: the Gviz track  </li>
<li><code>region</code>   : the region  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the dumped track  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>params</code>: use <code>displayPars(UcscTrack(genome="mm9", chromosome="chrM", track="knownGene"))</code> to see all available args.  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://rdrr.io/bioc/Gviz">r-Gviz</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pGenomePlot</p>
<ul>
<li>
<p><strong>description</strong><br />
    plot the genomic features</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>trkfiles:files</code>: the list of track dumped files  </li>
<li><code>region</code>: the region, in format of <code>chr1:1-1000</code>  </li>
<li><code>highlight</code>: the highlight regions, informat of start1-end1; start2-end2; ...  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the figure  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>genome</code>  : The genome  </li>
<li><code>showIdeo</code>: Show ideogram track? Default: True  </li>
<li><code>showAxis</code>: Show axis? Default: True  </li>
<li><code>showGenes</code>: Show geneTrack? Default: True  </li>
<li><code>params</code>: The params  <ul>
<li><code>genneral</code>:  General params for plotTracks</li>
<li><code>geneTrack</code>: The params for geneTrack</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://rdrr.io/bioc/Gviz">r-Gviz</a></p>
</li>
</ul>
</div>
<h3 id="gsea">gsea</h3>
<div class="admonition hint">
<p class="admonition-title">pGMT2Mat</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert a GMT file to a matrix.
    Rownames of GMT file will be the column names of output matrix.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file in GMT format.  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: output matrix file  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pExpmat2Gct</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert expression matrix to GCT file.
    Refer to http://software.broadinstitute.org/cancer/software/genepattern/file-formats-guide#GCT for file format</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>expfile:file</code>: the input expression matrix file. Samples as columns, genes as rows.  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the gct file  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSampleinfo2Cls</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert sample infomation to cls file.
    Refer to http://software.broadinstitute.org/cancer/software/genepattern/file-formats-guide#CLS for file format
    NOTE that the order of samples must be the same as in GMT file in further analysis.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>sifile:file</code>: the sample information file.  <ul>
<li>Headers are: [Sample, ]Patient, Group, Batch</li>
<li>Rows are samples</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the cls file  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSSGSEA</p>
<ul>
<li>
<p><strong>description</strong><br />
    Single sample GSEA
    Refer to http://software.broadinstitute.org/cancer/software/genepattern/file-formats-guide#GCT for GCT file format
    Refer to http://software.broadinstitute.org/cancer/software/genepattern/file-formats-guide#GMT for GMT file format</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>gctfile:file</code>: the expression file  </li>
<li><code>gmtfile:file</code>: the gmtfile for gene sets  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:file</code>: the output directory  </li>
<li><code>report.txt</code>: the enrichment report for each Gene set.</li>
<li><code>RES_&lt;GeneSet&gt;.png</code>: the running ES plot for <GeneSet></li>
<li><code>normP_&lt;GeneSet&gt;.png</code>: the norminal P value plot for <GeneSet></li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>weightexp</code>: Exponential weight employed in calculation of enrichment scores. Default: 0.75  </li>
<li><code>nperm</code>: Number of permutations. Default: 10000  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pGSEA</p>
<ul>
<li>
<p><strong>description</strong><br />
    GSEA
    Refer to http://software.broadinstitute.org/cancer/software/genepattern/file-formats-guide#GCT for GCT file format
    Refer to http://software.broadinstitute.org/cancer/software/genepattern/file-formats-guide#GMT for GMT file format
    Refer to http://software.broadinstitute.org/cancer/software/genepattern/file-formats-guide#CLS for CLS file format</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>gctfile:file</code>: the expression file  </li>
<li><code>clsfile:file</code>: the class file  </li>
<li><code>gmtfile:file</code>: the gmtfile for gene sets  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:file</code>: the output directory  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>weightexp</code>: Exponential weight employed in calculation of enrichment scores. Default: 0.75  </li>
<li><code>nperm</code>: Number of permutations. Default: 10000  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pEnrichr</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use APIs from http://amp.pharm.mssm.edu/Enrichr/help#api&amp;q=1 to analyze a gene list</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The gene list, each per line  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory, containing the tables and figures.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>topn</code>: Top N pathways used to plot. Default: 10  </li>
<li><code>col</code>: The columns index containing the genes. Default: 0  </li>
<li><code>delimit</code>: The delimit of input file. Default: '\t'  </li>
<li><code>dbs</code>: The databases to do enrichment against. Default: KEGG_2016  </li>
<li>A full list can be found here: http://amp.pharm.mssm.edu/Enrichr/#stats</li>
<li>Multiple dbs separated by comma (,)</li>
<li><code>norm</code>: Normalize the gene list use <a href="https://pypi.python.org/pypi/mygene/3.0.0">python-mygene</a>  </li>
<li><code>rmtags</code>: Remove pathway tags in the plot. Default: True  </li>
<li>For example: change "Lysine degradation_Homo sapiens_hsa00310" to "Lysine degradation".</li>
<li><code>plot</code>: Whether to plot the result. Default: True  </li>
<li><code>title</code>: The title for the plot. Default: "Gene enrichment: {db}"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://pypi.python.org/pypi/mygene/3.0.0">python-mygene</a> if <code>args.norm</code> is <code>True</code></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pTargetEnrichr</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use APIs from http://amp.pharm.mssm.edu/Enrichr/help#api&amp;q=1 to analyze a gene list</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The target genes with regulators  <ul>
<li>Format:</li>
<li>Header is not required, but may specified in first line starting with <code>#</code></li>
<li>If only 3 columns are there, the 3rd column is anyway the relation!</li>
<li>If only 4 columns are there, 3rd is target status, 4th is relation!
  <div class="highlight"><pre><span></span>#Regulator	Target	Regulator status	Target status	Relation
has-mir-22	Gene	+	+	+
</pre></div></li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory, containing the tables and figures.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>dbs</code>       : The databases to do enrichment against. Default: KEGG_2016  </li>
<li>A full list can be found here: http://amp.pharm.mssm.edu/Enrichr/#stats</li>
<li>Multiple dbs separated by comma (,)</li>
<li><code>rmtags</code>    : Remove pathway tags in the plot. Default: True  </li>
<li>For example: change "Lysine degradation_Homo sapiens_hsa00310" to "Lysine degradation".</li>
<li><code>enrplot</code>   : Whether to plot the result. Default: True  </li>
<li><code>enrn</code>      : Top N pathways used to plot. Default: 10  </li>
<li><code>netplot</code>   : Whether to plot the network. Default: True  </li>
<li><code>netn</code>      : Top N pathways used to plot the network. Default: 5  <ul>
<li>Must &lt;= <code>enrn</code>. If <code>netn</code> &gt;= <code>enrn</code>, <code>netn</code> = <code>enrn</code></li>
</ul>
</li>
<li><code>title</code>     : The title for the plot. Default: "Gene enrichment: {db}"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://pypi.python.org/pypi/mygene/3.0.0"><code>python-mygene</code></a>
    <a href="https://pypi.python.org/pypi/graphviz"><code>graphviz</code></a></p>
</li>
</ul>
</div>
<h3 id="hic">hic</h3>
<h3 id="marray">marray</h3>
<div class="admonition hint">
<p class="admonition-title">pCELdir2Matrix</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert CEL files to expression matrix
    File names will be used as sample names (colnames)</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: the directory containing the CEL files, could be gzipped  <ul>
<li>If you have files, then use <code>pFiles2Dir</code> first</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the expression matrix file  </li>
<li><code>outdir:dir</code>: the directory containing expr file and plots  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>pattern</code>  : The pattern to filter files. Default <code>'*'</code>  </li>
<li><code>norm</code>     : The normalization method. Default: rma (mas5)  </li>
<li><code>gfile</code>    : The group file. Default: ''  </li>
<li><code>cdffile</code>  : The cdffile. Default: ''  </li>
<li><code>annofile</code> : The annotation file. Default: ''  </li>
<li><code>hmrows</code>   : How many rows to be used to plot heatmap  </li>
<li><code>plot</code>: Whether to plot  <ul>
<li><code>boxplot</code>   : Whether to plot a boxplot. Default: False</li>
<li><code>heatmap</code>   : Whether to plot a heatmap. Default: False</li>
<li><code>histogram</code> : Whether to plot a histgram. Default: False</li>
</ul>
</li>
<li><code>devpars</code>    : Parameters for png. Default: <code>{'res': 300, 'width': 2000, 'height': 2000}</code>  </li>
<li><code>ggs</code>: The ggplot parameters  <ul>
<li><code>boxplot</code>  : The ggplot parameters for boxplot. Default: <code>Box(ylab = {0: "Log2 Intensity"})</code></li>
<li><code>heatmap</code>  : The ggplot parameters for heatmap. Default: <code>Box(theme = {'axis.text.y': 'r:element_blank()'})</code></li>
<li><code>histogram</code>: The ggplot parameters for histgram. Default: <code>Box(labs = {'x': "Log2 Intensity", "y": "Density"})</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<h3 id="misc">misc</h3>
<div class="admonition hint">
<p class="admonition-title">pGEP70</p>
<ul>
<li><strong>description</strong><br />
    Calculate GEP70 scores for multiple mylenoma 70-gene-signatures</li>
</ul>
</div>
<h3 id="pca">pca</h3>
<div class="admonition hint">
<p class="admonition-title">pPCA</p>
<ul>
<li>
<p><strong>description</strong><br />
    Perform PCA analysis</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The matrix to do the analysis  </li>
<li>Note that rows are samples, columns are features, if not, use <code>args.transpose = True</code></li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output coordinate file  </li>
<li>Columns are PCs, rows are samples</li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>transpose</code>: Whether to transpose the input matrix from infile. Default: False  </li>
<li><code>rownames</code>: The <code>row.names</code> argument for <code>read.table</code>, default: 1  </li>
<li><code>header</code>: The <code>header</code> argument for <code>read.table</code> to read the input file, default: True.  </li>
<li><code>screeplot</code>: Whether to generate the screeplot or not. Default: True  </li>
<li><code>sp_ncp</code>: Number of components in screeplot. Default: 0 (auto detect)  </li>
<li>if total # components (tcp) &lt; 20: use all</li>
<li>else if tcp &gt; 20, use 20</li>
<li><code>varplot</code>: Whether to generate the variable plot or not. Default: False  </li>
<li><code>biplot</code>: Whether to generate the variable plot or not. Default: True  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://cran.r-project.org/web/packages/factoextra/index.html"><code>r-factoextra</code></a> for plots</p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSelectPCs</p>
<ul>
<li>
<p><strong>description</strong><br />
    Select a subset of PCs from pPCA results</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: The directory generated from pPCA  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The file containing selected PCs  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>n</code>: The number of PCs to select. Default: 0.9  </li>
<li>If it is &lt; 1, used as the % variation explained from stdev.txt</li>
</ul>
</li>
</ul>
</div>
<h3 id="picard">picard</h3>
<div class="admonition hint">
<p class="admonition-title">pMarkDuplicates</p>
<ul>
<li>
<p><strong>description</strong><br />
    Identifies duplicate reads.</p>
<p>This tool locates and tags duplicate reads in a BAM or SAM file, where duplicate reads are defined as originating from a single fragment of DNA. Duplicates can arise during sample preparation e.g. library construction using PCR. See also EstimateLibraryComplexity for additional notes on PCR duplication artifacts. Duplicate reads can also result from a single amplification cluster, incorrectly detected as multiple clusters by the optical sensor of the sequencing instrument. These duplication artifacts are referred to as optical duplicates.</p>
<p>The MarkDuplicates tool works by comparing sequences in the 5 prime positions of both reads and read-pairs in a SAM/BAM file. An BARCODE_TAG option is available to facilitate duplicate marking using molecular barcodes. After duplicate reads are collected, the tool differentiates the primary and duplicate reads using an algorithm that ranks reads by the sums of their base-quality scores (default method).</p>
<p>The tool's main output is a new SAM or BAM file, in which duplicates have been identified in the SAM flags field for each read. Duplicates are marked with the hexadecimal value of 0x0400, which corresponds to a decimal value of 1024. If you are not familiar with this type of annotation, please see the following <a href="https://www.broadinstitute.org/gatk/blog?id=7019">blog post</a> for additional information.</p>
<p>Although the bitwise flag annotation indicates whether a read was marked as a duplicate, it does not identify the type of duplicate. To do this, a new tag called the duplicate type (DT) tag was recently added as an optional output in the 'optional field' section of a SAM/BAM file. Invoking the TAGGING_POLICY option, you can instruct the program to mark all the duplicates (All), only the optical duplicates (OpticalOnly), or no duplicates (DontTag). The records within the output of a SAM/BAM file will have values for the 'DT' tag (depending on the invoked TAGGING_POLICY), as either library/PCR-generated duplicates (LB), or sequencing-platform artifact duplicates (SQ). This tool uses the READ_NAME_REGEX and the OPTICAL_DUPLICATE_PIXEL_DISTANCE options as the primary methods to identify and differentiate duplicate types. Set READ_NAME_REGEX to null to skip optical duplicate detection, e.g. for RNA-seq or other data where duplicate sets are extremely large and estimating library complexity is not an aim. Note that without optical duplicate counts, library size estimation will be inaccurate.</p>
<p>MarkDuplicates also produces a metrics file indicating the numbers of duplicates for both single- and paired-end reads.</p>
<p>The program can take either coordinate-sorted or query-sorted inputs, however the behavior is slightly different. When the input is coordinate-sorted, unmapped mates of mapped records and supplementary/secondary alignments are not marked as duplicates. However, when the input is query-sorted (actually query-grouped), then unmapped mates and secondary/supplementary reads are not excluded from the duplication test and can be marked as duplicate reads.</p>
<p>If desired, duplicates can be removed using the REMOVE_DUPLICATE and REMOVE_SEQUENCING_DUPLICATES options.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The bam file   </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The marked bam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>params</code>: Other parameters for picard MarkDuplicates, default: ""  </li>
<li><code>tmpdir</code>: The tmpdir to use. Default: /tmp  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://broadinstitute.github.io/picard/">picard</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pAddOrReplaceReadGroups</p>
<ul>
<li>
<p><strong>description</strong><br />
    Replace read groups in a BAM file.This tool enables the user to replace all read groups in the INPUT file with a single new read group and assign all reads to this read group in the OUTPUT BAM file.</p>
<p>For more information about read groups, see the <a href="https://www.broadinstitute.org/gatk/guide/article?id=6472">GATK Dictionary entry</a>. </p>
<p>This tool accepts INPUT BAM and SAM files or URLs from the Global Alliance for Genomics and Health (GA4GH) (see http://ga4gh.org/#/documentation).</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The bam file  </li>
<li><code>rg</code>: The read group information. For example:  <ul>
<li>"RGID=4 RGLB=lib1 RGPL=illumina RGPU=unit1 RGSM=20"</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The bam file with read group added  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>picard</code>: The picard executable, default: "picard "  </li>
<li><code>params</code>: Other parameters for picard AddOrReplaceReadGroups, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://broadinstitute.github.io/picard/">picard</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCreateSequenceDictionary</p>
<ul>
<li>
<p><strong>description</strong><br />
    Creates a sequence dictionary for a reference sequence. This tool creates a sequence dictionary file (with ".dict" extension) from a reference sequence provided in FASTA format, which is required by many processing and analysis tools. The output file contains a header but no SAMRecords, and the header contains only sequence records.</p>
<p>The reference sequence can be gzipped (both .fasta and .fasta.gz are supported).</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The fasta file   </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The same fasta file, but with dict file created  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>params</code>: Other parameters for picard CreateSequenceDictionary, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://broadinstitute.github.io/picard/">picard</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCollectWgsMetrics</p>
<ul>
<li>
<p><strong>description</strong><br />
    Collect metrics about coverage and performance of whole genome sequencing (WGS) experiments.</p>
<p>This tool collects metrics about the fractions of reads that pass base- and mapping-quality filters as well as coverage (read-depth) levels for WGS analyses. Both minimum base- and mapping-quality values as well as the maximum read depths (coverage cap) are user defined.</p>
<p>Note: Metrics labeled as percentages are actually expressed as fractions!</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The bam file   </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The metrics file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>params</code>: Other parameters for <code>picard CollectWgsMetrics</code>, default: ""  </li>
<li><code>reffile</code>: The reference file, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://broadinstitute.github.io/picard/">picard</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSortSam</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use <code>picard SortSam</code> to sort sam or bam file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The sam or bam file to be sorted  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The sorted sam or bam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>order</code>: The sort order, default: coordinate. Possible: unsorted, queryname, coordinate, duplicate  </li>
<li><code>outtype</code>: The type of output file, sam or bam. Default: bam  </li>
<li><code>params</code>: Other parameters for <code>picard SortSam</code>, default: ""  </li>
<li><code>tmpdir</code>: The tmpdir to use. Default: /tmp  </li>
<li><code>javamem</code>: The memory for java vm. Default: "-Xms1g -Xmx8g"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://broadinstitute.github.io/picard/command-line-overview.html">picard</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pIndexBam</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use <code>picard BuildBamIndex</code> to index bam file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The bam file   </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The same bam file (link) but with .bai file in <code>proc.outdir</code>  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>picard</code>: The picard executable, default: "picard"  </li>
<li><code>params</code>: Other parameters for <code>picard BuildBamIndex</code>, default: "-Xms1g -Xmx8g"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://broadinstitute.github.io/picard/command-line-overview.html">picard</a></p>
</li>
</ul>
</div>
<h3 id="plot">plot</h3>
<div class="admonition hint">
<p class="admonition-title">pPlot</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use ggplot2 to generate plots</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input data file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnames</code> : Whether the input file has colnames. Default: True  </li>
<li><code>rnames</code> : Whether the input file has rownames. Default: False  </li>
<li><code>aes</code>    : The default aes. Default: {'x':1, 'y':2} (corresponding to colnames)  </li>
<li><code>helper</code> : Some helper codes to generate <code>params</code> and <code>ggs</code>  </li>
<li><code>devpars</code>: The device parameters. Default: <code>Box(res = 300, height = 2000, width = 2000)</code>  </li>
<li><code>ggs</code>    : The extra ggplot elements.  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pScatter</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use ggplot2 geom_point to generate plots</p>
</li>
<li>
<p><strong>infile</strong>  </p>
<ul>
<li><code>infile:file</code>: The input data file  </li>
</ul>
</li>
<li>
<p><strong>outfile</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnames</code> : Whether the input file has colnames. Default: True  </li>
<li><code>rnames</code> : Whether the input file has rownames. Default: False  </li>
<li><code>x</code>      : The x aes. Default: 1 (corresponding to colnames)  </li>
<li><code>y</code>      : The y aes. Default: 2 (corresponding to colnames)  </li>
<li><code>helper</code> : Some helper codes to generate <code>params</code> and <code>ggs</code>  </li>
<li><code>devpars</code>: The device parameters. Default: <code>Box(res = 300, height = 2000, width = 2000)</code>  </li>
<li><code>params</code> : The extra params for <code>geom_point</code>  </li>
<li><code>ggs</code>    : The extra ggplot elements.  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pPoints</p>
<ul>
<li><strong>description</strong><br />
    Alias for pScatter</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pHisto</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use ggplot2 geom_histogram to generate histograms</p>
</li>
<li>
<p><strong>infile</strong>  </p>
<ul>
<li><code>infile:file</code>: The input data file  </li>
</ul>
</li>
<li>
<p><strong>outfile</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnames</code> : Whether the input file has colnames. Default: True  </li>
<li><code>rnames</code> : Whether the input file has rownames. Default: False  </li>
<li><code>x</code>      : The x aes. Default: 1 (corresponding to colnames)  </li>
<li><code>helper</code> : Some helper codes to generate <code>params</code> and <code>ggs</code>  </li>
<li><code>devpars</code>: The device parameters. Default: <code>Box(res = 300, height = 2000, width = 2000)</code>  </li>
<li><code>params</code> : The extra params for <code>geom_point</code>  </li>
<li><code>ggs</code>    : The extra ggplot elements.  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pFreqpoly</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use ggplot2 geom_freqpoly to generate frequency polygon plot.</p>
</li>
<li>
<p><strong>infile</strong>  </p>
<ul>
<li><code>infile:file</code>: The input data file  </li>
</ul>
</li>
<li>
<p><strong>outfile</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnames</code> : Whether the input file has colnames. Default: True  </li>
<li><code>rnames</code> : Whether the input file has rownames. Default: False  </li>
<li><code>x</code>      : The x aes. Default: 1 (corresponding to colnames)  </li>
<li><code>helper</code> : Some helper codes to generate <code>params</code> and <code>ggs</code>  </li>
<li><code>devpars</code>: The device parameters. Default: <code>Box(res = 300, height = 2000, width = 2000)</code>  </li>
<li><code>params</code> : The extra params for <code>geom_point</code>  </li>
<li><code>ggs</code>    : The extra ggplot elements.  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBoxplot</p>
<ul>
<li>
<p><strong>description</strong><br />
    Generate box plot</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>datafile:file</code>: The data file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outpng:file</code>: The output figure  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>inopts</code> : Input options to read the input file  <ul>
<li><code>cnames</code> :   Whether the input file has header. Default: <code>True</code></li>
<li><code>rnames</code> :   Whether the input file has row names. Default: <code>False</code></li>
<li><code>delimit</code>:   The seperator. Defualt: <code>\\t</code></li>
</ul>
</li>
<li><code>x</code>      : The <code>ind</code> (index) column. Only for <code>args.stacked = True</code>. Default: <code>2</code>  </li>
<li><code>y</code>      : The <code>values</code> column. Only for <code>args.stacked = True</code>. Default: <code>1</code>  </li>
<li><code>helper</code> : Some raw codes to help to construct the matrix and arguments.  </li>
<li><code>stacked</code>: Whether the input file is stacked  <ul>
<li>Stacked file looks like:
  <div class="highlight"><pre><span></span>values	ind
1.1	col1
1.2	col1
...
.8	col2
.9	col2
...
3.2	col3
...
</pre></div></li>
<li>Unstacked file looks like:
  <div class="highlight"><pre><span></span>col1	col2	col3
1.1	.8	3.2
1.2	.9	2.2
</pre></div></li>
</ul>
</li>
<li><code>params</code>: Other parameters for <code>boxplot</code>, default: <code>""</code>  </li>
<li><code>ggs</code>   : Extra ggplot2 statements  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pHeatmap</p>
<ul>
<li>
<p><strong>description</strong><br />
    Plot heatmaps.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input matrix file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The heatmap  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>ggs</code>: The ggplot items for heatmap  </li>
<li><code>devpars</code>: The parameters for device. Default: <code>{'res': 300, 'height': 2000, 'width': 2000}</code>  </li>
<li><code>dendro</code>: The parameters for control of the dendrogram. Default: <code>{'dendro': True}</code>  <ul>
<li><code>dendro</code>: <code>True</code>: plot dendros for both rows and cols; <code>col</code>: only plot dendro for cols; <code>row</code>: only plot dendro for rows</li>
<li><code>rows</code>: The rownames to subset the rows and control the order of rows. Must a list. Only works when not plotting dendro for rows.</li>
<li><code>cols</code>: The colnames to subset the cols and control the order of cols. Must a list. Only works when not plotting dendro for cols.</li>
</ul>
</li>
<li><code>header</code>: The input file has header? Default: True  </li>
<li><code>rownames</code>: The input file has rownames? Default: 1  </li>
<li><code>rows</code>: Row selector  <ul>
<li><code>all</code>: All rows</li>
<li><code>top:N</code>: Top N rows (original data ordered in descending order). N defaults to 100</li>
<li><code>bottom:N</code>: Bottom N rows. N defaults to 100</li>
<li><code>both:N</code>: Top N rows and bottom N rows. N defaults to 50</li>
<li><code>random:N</code>: Random N rows. N defaults to 50</li>
<li><code>random-both:N</code>: Random N rows from top part and N rows from bottom part. N defaults to 50</li>
</ul>
</li>
<li><code>cols</code>: Col selector (see <code>rows</code>).  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pScatterCompare</p>
<ul>
<li>
<p><strong>description</strong><br />
    Plot scatter plot to compare values of first 2 columns of input data</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file containing a matrix with at least 2 columns  <ul>
<li>Other columns are groups used to group the scatter points</li>
<li>Data must be normalized to [0, 1]</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output plot  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>ggs</code>: Extra expressions for ggplot. Note if geom_point is included, original geom_point will be ignored.  </li>
<li><code>devpars</code>: The parameters for plot device. Default: <code>{'res': 300, 'height': 2000, 'width': 2000}</code>  </li>
<li><code>rownames</code>: Whether the input file has row names. Default: True  </li>
<li><code>regr</code>: Whether draw the regression line. Default: False  </li>
<li><code>corr</code>: The method to calculate the correlation. Default: <code>pearson</code>  <ul>
<li>Could be: <code>pearson</code>, <code>spearman</code> or <code>kendall</code></li>
<li>If it's neither of the three, no correlations will show.</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pROC</p>
<ul>
<li>
<p><strong>description</strong><br />
    Generate ROC curves and output AUC.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input matrix file.  <ul>
<li>Col1: rownames if args.rnames is True else label (0, 1 class)</li>
<li>Col2: prediction values from model1</li>
<li>...</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pVenn</p>
<ul>
<li>
<p><strong>description</strong><br />
    Venn/UpsetR plots.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input matrix  <ul>
<li>format:
<div class="highlight"><pre><span></span>	category1	category2	category3
[e1]	0	1	1
[e2]	0	0	1
...
[eN]	1	0	0
</pre></div>
rownames are not necessary but colnames are.</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The plot  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>    : Which tools to use. Default: auto (venn, upsetr, auto(n&lt;=3: venn, otherwise upsetr))  </li>
<li><code>rnames</code>  : Whether input file has rownames. Default: False  </li>
<li><code>params</code>  : Other params for <code>venn.diagram</code> or <code>upset</code>. Default: {}  </li>
<li><code>devpars</code> : The parameters for plot device. Default: <code>{'res': 300, 'height': 2000, 'width': 2000}</code>  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://www.rdocumentation.org/packages/VennDiagram"><code>r-VennDiagram</code></a>
    <a href="https://www.rdocumentation.org/packages/UpSetR"><code>r-UpSetR</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pPie</p>
<ul>
<li>
<p><strong>description</strong><br />
    Plot piechart</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file. Could be either:  <ul>
<li>Direct numbers of each category.
<div class="highlight"><pre><span></span>Group1	Group2
50	50
</pre></div></li>
<li>Presence of each items in the category.
<div class="highlight"><pre><span></span>	Group1	Group2
Item1	1	0
Item2	0	1
...
</pre></div></li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the output plot  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>rnames</code> : Whether the input file has row names. Default: <code>False</code>  </li>
<li><code>ggs</code>    : Extra expressions for ggplot.  </li>
<li><code>devpars</code>: The parameters for plot device. Default: <code>{'res': 300, 'height': 2000, 'width': 2000}</code>  </li>
</ul>
</li>
</ul>
</div>
<h3 id="power">power</h3>
<div class="admonition hint">
<p class="admonition-title">pSurvivalPower</p>
<ul>
<li>
<p><strong>description</strong><br />
    Do power analysis for survival analysis.
    See http://www.sample-size.net/sample-size-survival-analysis/</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file, could be either:  <ul>
<li>detailed suvival data with [<code>patient</code>, ]<code>time</code>, <code>status</code>, <code>variable1</code>, <code>variable2</code>, ...; or</li>
<li>ratios with <code>variable</code>, <code>survrate1</code>, <code>survrate2</code>, <code>ssratio</code>, where <code>survrate1</code> and
    <code>survrate2</code> are survival rates in group1 and group2, respectively,
    and <code>ssratio</code> is sample size ratio in group1/group2</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file with columns:  <ul>
<li>Variable: the variable (including paired groups)</li>
<li>Alpha: the alpha value</li>
<li>Beta: the beta value (1-power)</li>
<li>SSize1: the sample size for group1</li>
<li>SSize2: the sample size for group2</li>
<li>Total: the total sample size</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<h3 id="rank">rank</h3>
<div class="admonition hint">
<p class="admonition-title">pRankProduct</p>
<ul>
<li>
<p><strong>description</strong><br />
    Calculate the rank product of a set of ranks. Refer to <a href="https://en.wikipedia.org/wiki/Rank_product">here</a></p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
<li>Format:
<div class="highlight"><pre><span></span>			Case1	Case2	...
Feature1	8.2  	10.1 	...
Feature2	2.3  	8.0  	...
...
</pre></div></li>
<li>Or instead of values, you can also have ranks in the input file:
<div class="highlight"><pre><span></span>			Rank1	Rank2	...
Feature1	2    	1    	...
Feature2	3    	2    	...
...
</pre></div></li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file with original ranks, rank products and p-value if required  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>informat</code>: The input format of the values. Whether they are real values (value) or ranks (rank). Default: value  </li>
<li><code>pval</code>: Whether to calculate the p-value or not. Default: True  </li>
<li><code>header</code>: Whether the input file has headers (rownames are required!). Default: True  </li>
<li><code>plot</code>: Number of rows to plot. Default: 0 (Don't plot)  </li>
<li><code>cex</code>: Font size for plotting. Default: 0.9  </li>
<li><code>cnheight</code>: Colname height. Default: 80  </li>
<li><code>rnwidth</code>: Rowname width. Default: 50  </li>
<li><code>width</code>: Width of the png file. Default: 2000  </li>
<li><code>height</code>: height of the png file. Default: 2000  </li>
</ul>
</li>
</ul>
</div>
<h3 id="resource">resource</h3>
<div class="admonition hint">
<p class="admonition-title">pTxt</p>
<ul>
<li>
<p><strong>description</strong><br />
    Download CSV format files.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>in</code>: The name of the resource  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cols</code>: Select the columns to keep. Default: '' (all cols)  </li>
<li><code>rowfilter</code>: Filter rows. For example, to filter out rows not start with 'Chr':  <ul>
<li><code>"lambda x: not x[0].startswith('Chr')"</code></li>
<li>Note that rowfilter applied before cols filter.</li>
</ul>
</li>
<li><code>urls</code>: Available resources and their urls.  </li>
<li><code>gz</code>: Whether to gzip the output file.  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://en.wikipedia.org/wiki/CURL"><code>curl</code></a></p>
</li>
</ul>
</div>
<h3 id="rnaseq">rnaseq</h3>
<div class="admonition hint">
<p class="admonition-title">pEXPRdir2Matrix</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert expression files to expression matrix
    File names will be used as sample names (colnames)
    Each gene and its expression per line.
    Suppose each expression file has the same rownames and in the same order.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: the directory containing the expression files, could be gzipped  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the expression matrix file  </li>
<li><code>outdir:dir</code>: the directory containing expr file and plots  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>pattern</code> : The pattern to filter files. Default <code>'*'</code>  </li>
<li><code>namefunc</code>: Transform filename (no extension) as column name. Default: "function(fn) fn"  </li>
<li><code>header</code>  : Whether each expression file contains header. Default: <code>False</code>  </li>
<li><code>exrows</code>  : Rows to be excluded, regular expression applied. Default: <code>["^Sample", "^Composite", "^__"]</code>  </li>
<li><code>boxplot</code> : Whether to plot a boxplot. Default: False  </li>
<li><code>heatmap</code> : Whether to plot a heatmap. Default: False  </li>
<li><code>histplot</code>: Whether to plot a histgram. Default: False  </li>
<li><code>devpars</code> : Parameters for png. Default: <code>{'res': 300, 'width': 2000, 'height': 2000}</code>  </li>
<li><code>boxplotggs</code>: The ggplot parameters for boxplot. Default: <code>['r:ylab("Expression")']</code>  <ul>
<li>See ggplot2 documentation.</li>
</ul>
</li>
<li><code>heatmapggs</code>: The ggplot parameters for heatmap. Default: <code>['r:theme(axis.text.y = element_blank())']</code>  </li>
<li><code>histplotggs</code>: The ggplot parameters for histgram. Default: <code>['r:labs(x = "Expression", y = "# Samples")']</code>  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBatchEffect</p>
<ul>
<li>
<p><strong>description</strong><br />
    Remove batch effect with sva-combat.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>expr:file</code>: The expression file, generated by pEXPRdir2Matrix  </li>
<li><code>batch:file</code>: The batch file defines samples and batches.  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the expression matrix file  </li>
<li><code>outdir:dir</code>: the directory containing expr file and plots  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>    : The tool used to remove batch effect. Default <code>'combat'</code>  </li>
<li><code>hmrows</code>  : How many rows to be used to plot heatmap  </li>
<li><code>plot</code>: Whether to plot  <ul>
<li><code>boxplot</code>   : Whether to plot a boxplot. Default: False</li>
<li><code>heatmap</code>   : Whether to plot a heatmap. Default: False</li>
<li><code>histogram</code> : Whether to plot a histgram. Default: False</li>
</ul>
</li>
<li><code>devpars</code>    : Parameters for png. Default: <code>{'res': 300, 'width': 2000, 'height': 2000}</code>  </li>
<li><code>ggs</code>: The ggplot parameters  <ul>
<li><code>boxplot</code>  : The ggplot parameters for boxplot. Default: <code>Box(ylab = {0: "Log2 Intensity"})</code></li>
<li><code>heatmap</code>  : The ggplot parameters for heatmap. Default: <code>Box(theme = {'axis.text.y': 'r:element_blank()'})</code></li>
<li><code>histogram</code>: The ggplot parameters for histgram. Default: <code>Box(labs = {'x': "Log2 Intensity", "y": "Density"})</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pRawCounts2</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert raw counts to another unit</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: the expression matrix  <ul>
<li>rows are genes, columns are samples</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the converted expression matrix  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>transpose</code>: transpose the input matrix? default: False  </li>
<li><code>log2</code>: whether to take log2? default: False  </li>
<li><code>unit</code>: convert to which unit? default: cpm (or rpkm, tmm)  </li>
<li><code>header</code>: whether input file has header? default: True  </li>
<li><code>rownames</code>: the index of the column as rownames. default: 1  </li>
<li><code>glenfile</code>: the gene length file, for RPKM  <ul>
<li>no head, row names are genes, have to be exact the same order and length as the rownames of infile</li>
</ul>
</li>
<li><code>boxplot</code> : Whether to plot a boxplot. Default: False  </li>
<li><code>heatmap</code> : Whether to plot a heatmap. Default: False  </li>
<li><code>histplot</code>: Whether to plot a histgram. Default: False  </li>
<li><code>devpars</code> : Parameters for png. Default: <code>{'res': 300, 'width': 2000, 'height': 2000}</code>  </li>
<li><code>boxplotggs</code>: The ggplot parameters for boxplot. Default: <code>['r:ylab("Expression")']</code>  <ul>
<li>See ggplot2 documentation.</li>
</ul>
</li>
<li><code>heatmapggs</code>: The ggplot parameters for heatmap. Default: <code>['r:theme(axis.text.y = element_blank())']</code>  </li>
<li><code>histplotggs</code>: The ggplot parameters for histgram. Default: <code>['r:labs(x = "Expression", y = "# Samples")']</code>  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://bioconductor.org/packages/release/bioc/html/edger.html">edgeR</a> if cpm or rpkm is chosen
    <a href="https://rdrr.io/rforge/coseq/man/transform_RNAseq.html">coseq</a> if tmm is chosen</p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">p2RawCounts</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert gene expression to raw counts.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The expression matrix file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
<li><code>outdir:dir</code>: The output directory, may contain the figures.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>unit</code>: The unit of input gene expression. Default: fpkm  <ul>
<li>Could also be rpkm, tpm</li>
</ul>
</li>
<li><code>nreads</code>     : Total reads approximately. Default: 30, 000, 000  </li>
<li><code>refgene</code>    : The refgene file for gene length.  </li>
<li><code>boxplot</code>    : Whether to plot boxplot after transformation. Default: False  </li>
<li><code>heatmap</code>    : Whether to plot heatmap after transformation. Default: False  </li>
<li><code>heatmapn</code>   : How many genes used to plot heatmap. Default: 500  </li>
<li><code>histplot</code>   : Whether to plot histgram after transformation. Default: False  </li>
<li><code>devpars</code>    : The device parameters for plotting. Default: <code>{'res': 300, 'width': 2000, 'height': 2000}</code>  </li>
<li><code>boxplotggs</code> : The ggplot statement for boxplot.  </li>
<li><code>heatmapggs</code> : The ggplot statement for heatmap.  </li>
<li><code>histplotggs</code>: The ggplot statement for histgram.  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pRNAseqDEG</p>
<ul>
<li>
<p><strong>description</strong><br />
    Detect DEGs for RNA-seq data</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>efile:file</code>: The expression matrix  </li>
<li><code>gfile:file</code>: The group information  <ul>
<li>Like:
<div class="highlight"><pre><span></span>Sample1	Group1
Sample2	Group1
Sample3	Group1
Sample4	group2
Sample5	group2
Sample6	group2
</pre></div></li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The DEG list  </li>
<li><code>outdir:file</code>: The output directory containing deg list and plots  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>      : the tool used to detect DEGs. Default: 'edger' (deseq2)  </li>
<li><code>filter</code>    : filter out low count records. Default: <code>"1,2"</code> (At least 2 samples have at least 2 reads)  </li>
<li><code>mdsplot</code>   : whether to plot the MDS plot, default : True  </li>
<li><code>volplot</code>   : whether to plot the volcano plot, default : True  </li>
<li><code>maplot</code>    : whether to plot MA plots within each group, default : False  </li>
<li><code>heatmap</code>   : whether to plot the heatmap using DEGs. Default : False  </li>
<li><code>heatmapn</code>  : How many genes to be used for heatmap. If <code>heatmapn</code>, the number will be <code>heatmapn * # DEGs</code>. Default: 100  </li>
<li><code>heatmapggs</code>: The ggplots options for heatmap. Default : []  </li>
<li><code>maplotggs</code> : The ggplots options for maplot. Default : []  </li>
<li><code>volplotggs</code>: The ggplots options for volplot. Default : []  </li>
<li><code>devpars</code>   : Parameters for png. Default: <code>{'res': 300, 'width': 2000, 'height': 2000}</code>  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCoexp</p>
<ul>
<li><strong>description</strong><br />
    Get co-expression of gene pairs in the expression matrix.</li>
</ul>
</div>
<h3 id="sambam">sambam</h3>
<div class="admonition hint">
<p class="admonition-title">pSam2Bam</p>
<ul>
<li>
<p><strong>description</strong><br />
    Deal with mapped sam/bam files, including sort, markdup, and/or index</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output bam file  </li>
<li><code>idxfile:file</code>: The index of the output bam file  </li>
<li>If args.index == False, it'll a link to outfile and should be never used</li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>             : The tool used to do the sort. Default: sambamba (picard|sambamba|biobambam|samtools)  </li>
<li><code>sambamba</code>         : The path of the sambamba. Default: sambamba  </li>
<li><code>picard</code>           : The path of the picard. Default: picard  </li>
<li><code>biobambam_bamsort</code>: The path of the biobambam's bamsort. Default: bamsort  </li>
<li><code>samtools</code>         : The path of the samtools. Default: samtools  </li>
<li><code>sort</code>             : Do sorting? Default: True  </li>
<li>If input is sam, tool is biobambam, this should be True</li>
<li><code>index</code>            : Do indexing? Default: True  </li>
<li><code>markdup</code>          : Do duplicates marking? Default: False  </li>
<li><code>rmdup</code> for samtools will be called</li>
<li><code>rmdup</code>            : Do duplicates removing? Default: False  </li>
<li><code>tmpdir</code>           : The tmp dir used to store tmp files. Default: <system default tmpdir>  </li>
<li><code>sortby</code>           : Sort by coordinate or queryname. Default: coordinate  </li>
<li><code>nthread</code>          : Default: 1  </li>
<li><code>informat</code>         : The format of input file. Default: <detect from extension> (sam|bam)  </li>
<li><code>params</code>           : Other parameters for <code>tool</code>. Defaut: ""  </li>
<li><code>mem</code>              : The max memory to use. Default: "16G"  </li>
<li>Unit could be G/g/M/m</li>
<li>Will be converted to -Xmx4G, and -Xms will be 1/8 of it</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://lomereiter.github.io/sambamba/docs/sambamba-view.html">sambamba</a> if <code>args.tool</code> == samtools or reference used but not indexed.
    <a href="https://broadinstitute.github.io/picard/command-line-overview.html">picard</a>
    <a href="https://github.com/gt1/biobambam2">biobambam</a>
    <a href="https://github.com/samtools/samtools">samtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBamMarkdup</p>
<ul>
<li>
<p><strong>description</strong><br />
    Mark/remove duplicates for bam files</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output bam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>             : The tool used to do the sort. Default: sambamba (picard|sambamba|biobambam|samtools|bamutil)  </li>
<li><code>sambamba</code>         : The path of sambamba. Default: sambamba  </li>
<li><code>picard</code>           : The path of picard. Default: picard  </li>
<li><code>biobambam_bamsort</code>: The path of biobambam's bamsort. Default: bamsort  </li>
<li><code>samtools</code>         : The path of samtools. Default: samtools  </li>
<li><code>bamutil</code>          : The path of bamutil. Default: bam  </li>
<li><code>rmdup</code>            : Do duplicates removing? Default: False  </li>
<li>Samtools will anyway remove the duplicates</li>
<li><code>tmpdir</code>           : The tmp dir used to store tmp files. Default: <system default tmpdir>  </li>
<li><code>nthread</code>          : Default: 1  </li>
<li>Not available for samtools and picard</li>
<li><code>params</code>           : Other parameters for <code>tool</code>. Defaut: ""  </li>
<li><code>mem</code>              : The max memory to use. Default: "16G"  </li>
<li>Unit could be G/g/M/m</li>
<li>Will be converted to -Xmx4G, and -Xms will be 1/8 of it</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://lomereiter.github.io/sambamba/docs/sambamba-view.html">sambamba</a>
    <a href="https://broadinstitute.github.io/picard/command-line-overview.html">picard</a>
    <a href="https://github.com/gt1/biobambam2">biobambam</a>
    <a href="https://github.com/samtools/samtools">samtools</a>
    <a href="http://genome.sph.umich.edu/wiki/BamUtil#Programs">bamutil</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBamRecal</p>
<ul>
<li>
<p><strong>description</strong><br />
    Recalibrate a bam file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The bam file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output bam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>                         : The tool used to recalibrate the bam file. Default: <code>gatk</code> (gatk|bamutil)  </li>
<li><code>gatk</code>                         : The path of gatk, including java path. Default: <code>gatk</code>  </li>
<li><code>samtools</code>                     : The path of samtools. Default: <code>samtools</code>  </li>
<li><code>bamutil</code>                      : The path of bamutil. Default: <code>bam</code>  </li>
<li><code>picard</code>                       : The path of picard. Default: <code>picard</code>  </li>
<li><code>paramsRealignerTargetCreator</code> : Other parameters for <code>gatk RealignerTargetCreator</code>. Defaut: ""  </li>
<li><code>paramsIndelRealigner</code>         : Other parameters for <code>gatk IndelRealigner</code>. Defaut: ""  </li>
<li><code>paramsBaseRecalibrator</code>       : Other parameters for <code>gatk BaseRecalibrator</code>. Defaut: ""  </li>
<li><code>paramsPrintReads</code>             : Other parameters for <code>gatk PrintReads</code>. Defaut: ""  </li>
<li><code>params</code>                       : Other parameters for <code>bam recab</code>. Default: ""  </li>
<li><code>mem</code>                          : The max memory to use. Default: "32G"  </li>
<li><code>knownSites</code>                   : The known polymorphic sites to mask out. Default: "" (Required for GATK)  </li>
<li><code>ref</code>                          : The reference file. Required.  </li>
<li>Will be converted to -Xmx4G, and -Xms will be 1/8 of it</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://software.broadinstitute.org/gatk">gatk</a>
    <a href="https://github.com/samtools/samtools">samtools</a> if <code>args.ref</code> is not indexed, or bamutil is used for bam index file generation.
    <a href="https://broadinstitute.github.io/picard/command-line-overview.html">picard</a> if <code>args.ref is not dicted.</code></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBamReadGroup</p>
<ul>
<li>
<p><strong>description</strong><br />
    Add or replace read groups of a bam file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The bam file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output bam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>                         : The tool used. Default: <code>picard</code> (picard|bamutil)  </li>
<li><code>picard</code>                       : The path of picard. Default: <code>picard</code>  </li>
<li><code>bamutil</code>                      : The path of bamutil. Default: <code>bam</code>  </li>
<li><code>rg</code>                           : The read group. Default: {'id': '', 'pl': 'Illumina', 'pu': 'unit1', 'lb': 'lib1', 'sm': ''}  </li>
<li><code>id</code> will be parsed from filename with "<em>LX</em>" in it if not given</li>
<li><code>sm</code> will be parsed from filename</li>
<li><code>params</code>                       : Other parameters for <code>tool</code>. Defaut: ""  </li>
<li><code>mem</code>                          : The max memory to use. Default: "4G"  </li>
<li>Will be converted to -Xmx4G, and -Xms will be 1/8 of it</li>
<li><code>tmpdir</code>                       : The temporary directory. Default: <system tmpdir>  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://lomereiter.github.io/sambamba/docs/sambamba-view.html">gatk</a>
    <a href="https://github.com/samtools/samtools">samtools</a> if <code>args.ref</code> is not indexed.
    <a href="https://broadinstitute.github.io/picard/command-line-overview.html">picard</a> if <code>args.ref is not dicted.</code></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBamReorder</p>
<ul>
<li>
<p><strong>description</strong><br />
    Reorder a sam/bam file by a given reference file using <code>picard ReorderSam</code></p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The sam/bam file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output bam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>picard</code>                       : The path of picard. Default: <code>picard</code>  </li>
<li><code>ref</code>                          : The reference file. Required  </li>
<li><code>params</code>                       : Other parameters for <code>picard ReorderSam</code>. Defaut: ""  </li>
<li><code>mem</code>                          : The max memory to use. Default: "4G"  </li>
<li>Will be converted to -Xmx4G, and -Xms will be 1/8 of it</li>
<li><code>tmpdir</code>                       : The temporary directory. Default: <system tmpdir>  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://broadinstitute.github.io/picard/command-line-overview.html">picard</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBamMerge</p>
<ul>
<li>
<p><strong>description</strong><br />
    Merges multiple SAM and/or BAM files (must be sorted by coordinate) into a single file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infiles:file</code>: Input sam/bam files to be merged  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The merged bam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>     : The tool used to merge. Default: bamutil (picard|samtools|sambamba)  </li>
<li><code>picard</code>   : The path of picard. Default: <code>picard</code>  </li>
<li><code>bamutil</code>  : The path of bamutil. Default: <code>bam</code>  </li>
<li><code>samtools</code> : The path of samtools. Default: <code>samtools</code>  </li>
<li><code>sambamba</code> : The path of sambamba. Default: <code>sambamba</code>  </li>
<li><code>params</code>   : Other parameters for <code>tool</code>. Defaut: ""  </li>
<li><code>mem</code>      : The max memory to use. Default: "4G"  </li>
<li>Will be converted to -Xmx4G, and -Xms will be 1/8 of it, just for picard</li>
<li><code>tmpdir</code>   : The temporary directory. Default: <system tmpdir>  </li>
<li><code>nthread</code>  : # threads to use. Default: 1  </li>
<li>For picard, if nthread&gt;1, USE_THREADING=true, otherwise USE_THREADING=false</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://broadinstitute.github.io/picard/command-line-overview.html">picard</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBam2Gmut</p>
<ul>
<li>
<p><strong>description</strong><br />
    Call germline (snps and indels) from a call-ready bam file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input bam file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The vcf file containing the mutations  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used to call mutations. Default: gatk (vardict, snvsniffer, platypus, strelka)  </li>
<li><code>gatk</code>: The path of gatk. Default: gatk  </li>
<li><code>vardict</code>: The path of vardict. Default: vardict  </li>
<li><code>snvsniffer</code>: The path of snvsniffer. Default: SNVSniffer  </li>
<li><code>samtools</code>: The path of samtools. Default: samtools (used to generate reference index)  </li>
<li><code>platypus</code>: The path of platypus. Default: platypus  </li>
<li><code>strelka</code>: The path of strelka. Default: configureStrelkaGermlineWorkflow.py  </li>
<li><code>configParams</code>: The params for <code>strelka</code> configuration. Default: ""  </li>
<li><code>picard</code>: The path of picard. Default: picard  </li>
<li><code>mem</code>: The memory to be used. Default: 32G  </li>
<li>will be converted to -Xms4G -Xmx32G for java programs</li>
<li><code>ref</code>: The reference file. Required.  </li>
<li><code>gz</code>: Gzip output file? Default: False  </li>
<li><code>tmpdir</code>: The temporary directory. Default: <system tmpdir>  </li>
<li><code>params</code>: Other params for <code>tool</code>. Default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://lomereiter.github.io/sambamba/docs/sambamba-view.html">gatk</a>
    <a href="https://github.com/samtools/samtools">samtools</a> if <code>args.ref</code> is not indexed.
    <a href="https://broadinstitute.github.io/picard/command-line-overview.html">picard</a> if <code>args.ref is not dicted.</code>
    <a href="https://github.com/AstraZeneca-NGS/VarDict">vardict</a>
    <a href="http://snvsniffer.sourceforge.net/homepage.htm#latest">snvsniffer</a>
    <a href="http://www.well.ox.ac.uk/platypus">platypus</a>
    <a href="https://github.com/Illumina/strelka">strelka@2.7.1+</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBamPair2Smut</p>
<ul>
<li>
<p><strong>description</strong><br />
    Call somatic mutations from tumor-normal bam pair.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>tumor:file</code>: The tumor bam file  </li>
<li><code>normal:file</code>: The normal bam file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The vcf file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used to call mutations. Default: gatk (somaticsniper, strelka, snvsniffer, virmid, varidct)  </li>
<li><code>gatk</code>: The path to gatk. Default: gatk  </li>
<li><code>somaticsniper</code>: The path to gatk. Default: bam-somaticsniper  </li>
<li><code>strelka</code>: The path to gatk. Default: configureStrelkaSomaticWorkflow.py  </li>
<li><code>snvsniffer</code>: The path to gatk. Default: SNVSniffer  </li>
<li><code>virmid</code>: The path to gatk. Default: virmid  </li>
<li><code>vardict</code>: The path to gatk. Default: vardict  </li>
<li><code>samtools</code>: The path to gatk. Default: samtools  </li>
<li><code>picard</code>: The path to gatk. Default: picard  </li>
<li><code>configParams</code>: The configuration parameters for <code>configureStrelkaSomaticWorkflow.py</code>. Default: <code>{}</code>  </li>
<li><code>params</code>: The parameters for main programs. Default: <code>{}</code>  </li>
<li><code>meme</code>: The memory. Default: 24G  </li>
<li><code>ref</code>: The reference genom. Default: <code>params.ref.value</code>  </li>
<li><code>gz</code>: Whether gzip the output vcf file. Default: False  </li>
<li><code>nthread</code>: The number of threads to use. Default: 1  </li>
<li><code>tmpdir</code>: The temporary directory. Default: <code>params.tmpdir.value</code>  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://lomereiter.github.io/sambamba/docs/sambamba-view.html">gatk</a>
    <a href="https://github.com/samtools/samtools">samtools</a> if <code>args.ref</code> is not indexed.
    <a href="https://broadinstitute.github.io/picard/command-line-overview.html">picard</a> if <code>args.ref is not dicted.</code>
    <a href="https://github.com/AstraZeneca-NGS/VarDict">vardict</a>
    <a href="http://snvsniffer.sourceforge.net/homepage.htm#latest">snvsniffer</a>
    <a href="http://www.well.ox.ac.uk/platypus">platypus</a>
    <a href="https://github.com/Illumina/strelka">strelka@2.7.1+</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBam2Cnv</p>
<ul>
<li>
<p><strong>description</strong><br />
    Detect copy number variation from bam files.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>input:file</code>: The bam file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output vcf file  </li>
<li><code>outdir</code>: The output directory containing other result files  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>gz</code>                    : Whether to gzip the output vcf file. Default: False  </li>
<li><code>tool</code>                  : The tool used to call cnv. Default: 'cnvkit'  </li>
<li><code>cnvnator</code>              : The path of cnvnator. Default: 'cnvnator'  </li>
<li><code>cnvnator2vcf</code>          : The path of cnvnator2VCF. Default: 'cnvnator2VCF.pl'  </li>
<li><code>cnvkit</code>                : The path of cnvkit. Default: 'cnvkit.py'  </li>
<li><code>wandy</code>                 : Tha path of Wandy. Default: 'Wandy'. A <code>tool.info</code> file should be with the executable file.  </li>
<li><code>ref</code>                   : The reference file. Required by cnvkit to generate access file. Default: ''  </li>
<li><code>cnvkitAccessParams</code>    : The params for cnvkit access command. Default: '-s 5000'  </li>
<li><code>cnvkitTargetParams</code>    : The params for cnvkit target command. Default: '--split --short-names'  </li>
<li><code>cnvkitCoverageParams</code>  : The params for cnvkit coverage command. Default: ''  </li>
<li><code>cnvkitReferenceParams</code> : The params for cnvkit reference command. Default: '--no-edge'  </li>
<li><code>cnvkitFixParams</code>       : The params for cnvkit fix command. Default: '--no-edge'  </li>
<li><code>cnvkitSegmentParams</code>   : The params for cnvkit segment command. Default: ''  </li>
<li><code>cnvkitCallParams</code>      : The params for cnvkit call command. Default: ''  </li>
<li><code>cnvkitPlotParams</code>      : The params for cnvkit plot command. Default: ''  </li>
<li><code>cnvkitBreaksParams</code>    : The params for cnvkit breaks command. Default: ''  </li>
<li><code>cnvkitGainlossParams</code>  : The params for cnvkit gainloss command. Default: ''  </li>
<li><code>cnvkitMetricsParams</code>   : The params for cnvkit metrics command. Default: ''  </li>
<li><code>cnvkitSegmetricsParams</code>: The params for cnvkit segmetrics command. Default: '--iqr'  </li>
<li><code>cnvkitExportParams</code>    : The params for cnvkit export command. Default: ''  </li>
<li><code>cnvkitScatterParams</code>   : The params for cnvkit scatter command. Default: [''] # multiple scatter plots  </li>
<li><code>cnvkitHeatmapParams</code>   : The params for cnvkit heatmap command. Default: [''] # multiple heatmap plots  </li>
<li><code>cnvkitDiagramParams</code>   : The params for cnvkit diagram command. Default: ''  </li>
<li><code>cnvkitReport</code>          : Generate cnvkit reports? Default: True  </li>
<li><code>cnvkitPlot</code>            : Generate cnvkit plots? Default: True  </li>
<li><code>cnvnatorBinsize</code>       : Bin size for cnvnator. Default: 100  </li>
<li><code>cnvnatorGenome</code>        : Genome for cnvnator. Default: 'hg19'. (NCBI36, hg18, GRCh37, hg19)  </li>
<li><code>params</code>                : The params for <code>tool</code>. Default: '-t 1' # wandy 1:hg19 solid cell/blood, 2:hg19 cell free/plamsa, 3:hg38 solid cell/blood, 4:hg38 cell free/plamsa  </li>
<li><code>mem</code>                   : The memory used. Default: '20G' # only for wandy  </li>
<li><code>nthread</code>               : The # threads to use. Default: 1     # only for cnvkit  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://cnvkit.readthedocs.io/en/stable/index.html"><code>cnvkit</code></a>
    <a href="https://github.com/abyzovlab/CNVnator"><code>cnvnator</code></a></p>
<ul>
<li><code>wandy</code>: Inside cnv caller  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBamStats</p>
<ul>
<li>
<p><strong>description</strong><br />
    Get read depth from bam files.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input bam file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output statistic file  </li>
<li><code>outdir:dir</code>: The directory containing result files and figures.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used to do the job. Default: bamstats  </li>
<li><code>bamstats</code>: The path to bamstats. Default: bamstats  </li>
<li><code>params</code>: Other params to main program. Default: <code>{}</code>  </li>
<li><code>mem</code>: The memory to be used. Default: 16G  </li>
<li><code>plot</code>: Whether plot the result. Default: True  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBam2Fastq</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert sam/bam files to pair-end fastq files.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The sam/bam file.  <ul>
<li>Sam files only available for biobambam, picard</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>fqfile1:file</code>: The 1st match of paired reads  </li>
<li><code>fqfile2:file</code>: The 2nd match of paired reads  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>     : The tool to use. Default: biobambam (bedtools, samtools, picard)  </li>
<li><code>biobambam</code>: The path of bamtofastq of biobambam. Default: bamtofastq  </li>
<li><code>bedtools</code> : The path of bedtools. Default: bedtools  </li>
<li><code>samtools</code> : The path of samtools. Default: samtools  </li>
<li><code>picard</code>   : The path of picard. Default: picard  </li>
<li><code>mem</code>      : The memory to be used by picard. Default: 8G  </li>
<li><code>gz</code>       : Whether gzip the output files. Default: True  </li>
<li><code>params</code>: : Other params for <code>tool</code>. Default: ''  </li>
<li><code>tmpdir</code>   : The tmpdir. Default: <code>__import__('tempfile').gettempdir()</code>  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://broadinstitute.github.io/picard/command-line-overview.html">picard</a>
    <a href="https://github.com/gt1/biobambam2">biobambam</a>
    <a href="https://github.com/samtools/samtools">samtools</a>
    <a href="http://bedtools.readthedocs.io/en/latest/content/bedtools-suite.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBam2FastqSE</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert sam/bam files to single-end fastq files.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The sam/bam file.  <ul>
<li>Sam files only available for biobambam, picard</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>fqfile:file</code>: The fastq file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>     : The tool to use. Default: biobambam (bedtools, samtools, picard)  </li>
<li><code>biobambam</code>: The path of bamtofastq of biobambam. Default: bamtofastq  </li>
<li><code>bedtools</code> : The path of bedtools. Default: bedtools  </li>
<li><code>samtools</code> : The path of samtools. Default: samtools  </li>
<li><code>picard</code>   : The path of picard. Default: picard  </li>
<li><code>mem</code>      : The memory to be used by picard. Default: 8G  </li>
<li><code>gz</code>       : Whether gzip the output files. Default: True  </li>
<li><code>params</code>: : Other params for <code>tool</code>. Default: ''  </li>
<li><code>tmpdir</code>   : The tmpdir. Default: <code>__import__('tempfile').gettempdir()</code>  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://broadinstitute.github.io/picard/command-line-overview.html">picard</a>
    <a href="https://github.com/gt1/biobambam2">biobambam</a>
    <a href="https://github.com/samtools/samtools">samtools</a>
    <a href="http://bedtools.readthedocs.io/en/latest/content/bedtools-suite.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pBam2Counts</p>
<ul>
<li>
<p><strong>description</strong><br />
    Extract read counts from RNA-seq bam files.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input bam files  </li>
</ul>
</li>
<li>
<p><strong>outfile</strong>  </p>
<ul>
<li><code>outfile:file</code>: The count file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used to extract counts. Default: ht-seq  </li>
<li><code>htseq</code>: The path of htseq-count.  </li>
<li><code>params</code>: Other params for main program.  </li>
<li><code>refgene</code>: The reference gene in GTF format.  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://htseq.readthedocs.io/"><code>htseq</code></a></p>
</li>
</ul>
</div>
<h3 id="seq">seq</h3>
<div class="admonition hint">
<p class="admonition-title">pConsvPerm</p>
<ul>
<li>
<p><strong>description</strong><br />
    Generate a null distribution of conservation scores.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>seed</code>: The seed to generate the random regions. Default: None  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: A file with mean conservation scores sorted descendingly.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>len</code>: The length of a random region. Default: 50  </li>
<li><code>nperm</code>: Number of permutations. Default: 1000  </li>
<li><code>gsize</code>: The chrom size file.  </li>
<li><code>bedtools</code>: The path of bedtools.  </li>
<li><code>bwtool</code>: The path of bwtool.  </li>
<li><code>consvdir</code>: The directory containing bigwig files of conservation scores  <ul>
<li>The bigwig file should start with chr name: chrN.*</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/CRG-Barcelona/bwtool">bwtool</a>
    <a href="http://bedtools.readthedocs.io/en/latest/content/bedtools-suite.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pConsv</p>
<ul>
<li>
<p><strong>description</strong><br />
    Get the conservation scores of regions.
    It uses wigFix to find the conservation scores.
    But first you have to convert those wigFix.gz files to bigWig files using ucsc-wigToBigWig</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>bedfile:file</code>: The bedfile with regions in the same chromosome  </li>
<li><code>permfile:file</code>: The permutaiton file generated by <code>pConsvPerm</code>, used to calculate p-values  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>consvdir</code>: The bigwig directory, the bigwig files must be named as "chrN.*.bw"  <ul>
<li>For example: <code>chr1.phyloP30way.bw</code></li>
</ul>
</li>
<li><code>bwtool</code>: The path of bwtool executable. Default: <code>bwtool</code>  </li>
<li><code>bedtools</code>: The path of bedtools executable. Default: <code>bedtools</code>  </li>
<li><code>pval</code>: Whether calculate pvalue of each region. Default: False  <ul>
<li>In this case, the <code>in.permfile</code> can be ignored.</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/CRG-Barcelona/bwtool">bwtool</a>
    <a href="http://bedtools.readthedocs.io/en/latest/content/bedtools-suite.html">bedtools</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pPromoters</p>
<ul>
<li>
<p><strong>description</strong><br />
    Get the promoter regions in bed format of a gene list give in infile.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: the gene list file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the bed file containing the promoter region  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>up</code>: the upstream to the tss, default: 2000  </li>
<li><code>down</code>: the downstream to the tss, default: 2000  </li>
<li><code>genome</code>: the genome, default: hg19  </li>
</ul>
</li>
<li>
<p><strong>require</strong><br />
<a href="http://mygene.info/">python-mygene</a></p>
</li>
</ul>
</div>
<h3 id="snp">snp</h3>
<div class="admonition hint">
<p class="admonition-title">pSnp2Bedx</p>
<ul>
<li>
<p><strong>description</strong><br />
    Find coordinates for SNPs in BEDX format.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>snpfile:file</code>: the snp file, each snp per line  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the result file, columns are:  <ul>
<li>chrom, start(0-based), end, name, score, strand, ref, allele</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>genome</code>: default: hg19  </li>
<li><code>snpver</code>: default: snp147  </li>
<li><code>notfound</code>: What to do if the snp is not found. Default: skip  </li>
<li><code>inmeta</code>: The metadata for input file to determine which column is rsID  </li>
<li><code>xcols</code>: The extra columns to extract and output to extra columns in output file.  </li>
<li><code>indem</code>: The input delimit. Default: '\t'  </li>
<li><code>incom</code>: The input comment. Default: '#'  </li>
<li><code>skip</code>: The lines to skip for input file. Default: 0  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/brentp/cruzdb"><code>python-cruzdb</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSnp2Avinput</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert SNP list to avinput to ANNOVAR.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>snpfile:file</code>: the snp file, each snp per line  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the result avinput file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>genome</code>: default: hg19  </li>
<li><code>snpver</code>: default: snp147  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/brentp/cruzdb"><code>python-cruzdb</code></a></p>
</li>
</ul>
</div>
<h3 id="snparray">snparray</h3>
<div class="admonition hint">
<p class="admonition-title">pGistic</p>
<ul>
<li>
<p><strong>description</strong><br />
    Runing GISTIC to get CNV results.
    see: ftp://ftp.broadinstitute.org/pub/GISTIC2.0/GISTICDocumentation_standalone.htm</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>segfile:file</code>: Segmentation File  </li>
<li><code>mkfile:file</code> : Markers File  </li>
<li><code>alfile:file</code> : Array List File  </li>
<li><code>cnvfile:file</code>: CNV File  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory  <ul>
<li>All Lesions File (all_lesions.conf_XX.txt, where XX is the confidence level)</li>
<li>Amplification Genes File (amp_genes.conf_XX.txt, where XX is the confidence level)</li>
<li>Deletion Genes File (del_genes.conf_XX.txt, where XX is the confidence level)</li>
<li>Gistic Scores File (scores.gistic)</li>
<li>Segmented Copy Number (raw_copy_number.pdf)</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>gistic</code>: The path to gistic.  </li>
<li><code>genome</code>: The genome used to select refgene file from refgenefiles.  </li>
<li><code>mcr</code>: The mcr path  </li>
<li><code>params</code>: Other params for gistic  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSNP6Genotype</p>
<ul>
<li>
<p><strong>description</strong><br />
    Call genotypes from GenomeWideSNP_6 CEL file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>celfile:file</code>: the CEL file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the outfile containing probe name and genotypes  </li>
<li>format: <code>&lt;Probe name&gt;\t&lt;genotype&gt;</code></li>
<li><code>&lt;genotype&gt;</code> = 0: AA, 1: AB, 2: BB</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bioconductor.org/packages/release/bioc/html/crlmm.html">bioconductor-crlmm</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pGenoToAvInput</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert the genotype called by pSNP6Genotype to <a href="http://annovar.openbioinformatics.org/en/latest/user-guide/input/#annovar-input-file">ANNOVAR input file</a> using dbSNP identifiers. </p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>genofile:file</code>: the genofile generated by pSNP6Genotype, must be sorted by probe names  </li>
<li><code>annofile:flie</code>: the annotation file downloaded from http://www.affymetrix.com/support/technical/annotationfilesmain.affx  <ul>
<li>Could be in .gz format</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the avinput file  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/pwwang/read2">python-read2</a></p>
</li>
</ul>
</div>
<h3 id="sql">sql</h3>
<div class="admonition hint">
<p class="admonition-title">pCreateTable</p>
<ul>
<li>
<p><strong>description</strong><br />
    Create tables in the database</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>dsn</code>: The dsn to connect to the database  <ul>
<li>currently support <code>sqlite:file=...</code></li>
</ul>
</li>
<li><code>schema:file</code>: The schema file  <ul>
<li>could be a pure schema file:
<div class="highlight"><pre><span></span>Field	Type	Statement
ID	INT	PRIMARY KEY
...
</pre></div></li>
<li>or a data file with header</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>dsn</code>: The dsn  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>intype</code>: The input file schema file or a data file. Default: <code>schema</code>  </li>
<li><code>drop</code>: Force creating the table (drop the pre-existing table)  </li>
<li><code>delimit</code>: The delimit of input file. Default: <code>\\t</code>  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pImportData</p>
<ul>
<li>
<p><strong>description</strong><br />
    Create tables and import the data</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>dsn</code>: The dsn to connect to the database  <ul>
<li>currently support <code>sqlite:file=...</code></li>
</ul>
</li>
<li><code>datafile:file</code>: The schema file  <ul>
<li>must have header</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>dsn</code>: The dsn  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>delimit</code>: The delimit of input file. Default: <code>\\t</code>  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pUpdateTable</p>
<ul>
<li>
<p><strong>description</strong><br />
    Update table using sql.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>dsn</code>: The dsn to connect to the database  <ul>
<li>currently support <code>sqlite:file=...</code></li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>dsn</code>: The dsn  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>sql</code>: The sql to update the table (list)  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSelectTable</p>
<ul>
<li>
<p><strong>description</strong><br />
    Select data from table and dump it.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>dsn</code>: The dsn to connect to the database  <ul>
<li>currently support <code>sqlite:file=...</code></li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The dumped file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>sql</code>: The sql to select data from the table (list)  </li>
</ul>
</li>
</ul>
</div>
<h3 id="stats">stats</h3>
<div class="admonition hint">
<p class="admonition-title">pMetaPval</p>
<ul>
<li>
<p><strong>description</strong><br />
    Combine p-values in the files from input directory</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:dir</code>: The directory containing the input files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file containing the meta-pvalues  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>args.pattern</code>: The pattern used to filter the input files. Default: '*'  </li>
<li><code>args.header</code>: Whether the input files contains a header. Default: True  <ul>
<li>Could be a list to specify it for each file.</li>
<li>The order should be concordant with the file names</li>
</ul>
</li>
<li><code>args.pcol</code>: Which column is the p-value. Default: -1 (last column)  </li>
<li><code>args.poutonly</code>: Only output pvalues. Default: False (output all possible information)  </li>
<li><code>args.outheader</code>: Whether output the header. Default: True  </li>
<li><code>args.method</code>: The method used to calculate the meta-pvalue. Default: sumlog (Fisher's method)  <ul>
<li>Other available methods: logitp, sumz, votep, sump, meanp and wilkinsonp</li>
<li>See: https://www.rdocumentation.org/packages/metap/versions/0.8</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://www.rdocumentation.org/packages/metap/"><code>r-matep</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMetaPval1</p>
<ul>
<li>
<p><strong>description</strong><br />
    Combine p-values in a single file by rows.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file containing the meta-pvalues  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>args.header</code>: Whether the input files contains a header. Default: True  </li>
<li><code>args.pcol</code>: Which column is the p-value. Default: -1 (last column)  </li>
<li><code>args.poutonly</code>: Only output pvalues. Default: False (output all possible information)  </li>
<li><code>args.outheader</code>: Whether output the header. Default: True  </li>
<li><code>args.method</code>: The method used to calculate the meta-pvalue. Default: sumlog (Fisher's method)  <ul>
<li>Other available methods: logitp, sumz, votep, sump, meanp and wilkinsonp</li>
<li>See: https://www.rdocumentation.org/packages/metap/versions/0.8</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://www.rdocumentation.org/packages/metap/"><code>r-matep</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSurvival</p>
<ul>
<li>
<p><strong>description</strong><br />
    Survival analysis</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file (header is required).  <ul>
<li>col1: rownames if args.inopts.rnames = True</li>
<li>col2: the survival time</li>
<li>col3: the status. 0/1 for alive/dead or 1/2 for alive dead</li>
<li>col4: var1.</li>
<li>... other variables</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The outfile containing the pvalues  </li>
<li><code>outdir:dir</code>  : The output directory containing the pval files and plots  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>inunit</code>    : The time unit in input file. Default: days  </li>
<li><code>outunit</code>   : The output unit for plots. Default: days  </li>
<li><code>nthread</code>   : Number of threads used to perform analysis for groups. Default: 1  </li>
<li><code>inopts</code>    : The options for input file  <ul>
<li><code>rnames</code>: Whether input file has row names. Default: True</li>
</ul>
</li>
<li><code>combine</code>   : Whether combine groups in the same plot. Default: True  </li>
<li><code>devpars</code>   : The device parameters for png. Default: <code>{res:300, height:2000, width:2000}</code>  <ul>
<li>The height and width are for each survival plot. If args.combine is True, the width and height will be multiplied by <code>max(arrange.ncol, arrange.nrow)</code></li>
</ul>
</li>
<li><code>covfile</code>   : The covariant file. Require rownames in both this file and input file.  </li>
<li><code>ngroups</code>   : Number of curves to plot (the continuous number will divided into <code>ngroups</code> groups.  </li>
<li><code>plot</code>      : The params for plot.  <ul>
<li><code>params</code> : The params for <code>ggsurvplot</code>. Default: <code>Box({'risk.table': True, 'conf.int': True, 'font.legend': 13, 'pval': '{method}\np = {pval}'})</code><ul>
<li>You may do <code>ylim.min</code> to set the min ylim. Or you can set it as 'auto'. Default: 0. </li>
</ul>
</li>
<li><code>arrange</code>: How to arrange multiple survival plots in one if <code>args.combine = True</code>.<ul>
<li><code>nrow</code>: The number of rows. Default: 1</li>
<li><code>ncol</code>: The number of cols. Default: 1</li>
</ul>
</li>
</ul>
</li>
<li><code>ggs</code>       : Extra ggplot2 elements for main plot. <code>ggs.table</code> is for the risk table.  </li>
<li><code>pval</code>      : The method to calculate the pvalue shown on the plot. Default: True (logrank)  <ul>
<li>Could also be <code>waldtest</code>, <code>likeratio</code> (Likelihoold ratio test)</li>
</ul>
</li>
<li><code>method</code>    : The method to do survival analysis.   </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://rdrr.io/cran/survival/"><code>r-survival</code></a>
    <a href="https://rdrr.io/cran/survminer/"><code>r-survminer</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pChiSquare</p>
<ul>
<li>
<p><strong>description</strong><br />
    Do chi-square test.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file.  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code> : The output file containing Xsquare, df, pval and method  </li>
<li><code>obsvfile:file</code>: The observation matrix  </li>
<li><code>exptfile:file</code>: The expectation matrix  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>intype</code>: The type of the input file:  <ul>
<li><code>count</code> (default): The contingency table
<div class="highlight"><pre><span></span>#         | Disease | Healthy |
# --------+---------+---------+
#   mut   |   40    |   12    |
# non-mut |   23    |   98    |
# --------+---------+---------+
</pre></div></li>
<li><code>raw</code>: The raw values:
<div class="highlight"><pre><span></span># Contingency table rows: Mut, Non
# Contingency table cols: Disease, Healthy
#
#         | S1 | S2 | ... | Sn |
# --------+----+----+-----+----+
# Disease | 1  | 0  | ... | 1  |
# Healthy | 0  | 1  | ... | 0  |
# --------+----+----+-----+----+
# Mut     | 1  | 0  | ... | 1  |
# Non     | 0  | 1  | ... | 0  |
</pre></div></li>
</ul>
</li>
<li><code>ctcols</code>: The colnames of contingency table if input file is raw values  <ul>
<li>You may also specify them in the head of the input file</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pFisherExact</p>
<ul>
<li>
<p><strong>description</strong><br />
    Do fisher exact test.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file.  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code> : The output file containing confInt1, confInt2, oddsRatio, pval, alternative and method.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>intype</code>: The type of the input file:  <ul>
<li><code>count</code> (default): The contingency table
<div class="highlight"><pre><span></span>#         | Disease | Healthy |
# --------+---------+---------+
#   mut   |   40    |   12    |
# non-mut |   23    |   98    |
# --------+---------+---------+
</pre></div></li>
<li><code>raw</code>: The raw values:
<div class="highlight"><pre><span></span># Contingency table rows: Mut, Non
# Contingency table cols: Disease, Healthy
#
#         | S1 | S2 | ... | Sn |
# --------+----+----+-----+----+
# Disease | 1  | 0  | ... | 1  |
# Healthy | 0  | 1  | ... | 0  |
# --------+----+----+-----+----+
# Mut     | 1  | 0  | ... | 1  |
# Non     | 0  | 1  | ... | 0  |
</pre></div></li>
</ul>
</li>
<li><code>ctcols</code>: The colnames of contingency table if input file is raw values  <ul>
<li>You may also specify them in the head of the input file</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pPWFisherExact</p>
<ul>
<li>
<p><strong>description</strong><br />
    Do pair-wise fisher exact test.
    Commonly used for co-occurrence/mutual-exclusivity analysis.
    P-value indicates if the pairs are significantly co-occurred or mutually exclusive.
    Co-occurrence: Odds ratio &gt; 1
    Mutual-exclusivity: Odds ratio &lt; 1</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file.  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code> : The output file containing confInt1, confInt2, oddsRatio, pval, qval, alternative and method.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>intype</code>: The type of the input file:  <ul>
<li><code>pairs</code>: The contingency table
<div class="highlight"><pre><span></span>#
# A+	B+	4
# A-	B-	175
# A+	B-	12
# A-	B+	1
#
</pre></div></li>
<li><code>raw</code> (default): The raw values:
<div class="highlight"><pre><span></span>#
#         | S1 | S2 | ... | Sn |
# --------+----+----+-----+----+
# A       | 1  | 0  | ... | 1  |
# B       | 0  | 1  | ... | 0  |
# ...     |           ...      |
# X       | 0  | 1  | ... | 0  |
# --------+----+----+-----+----+
#
</pre></div></li>
</ul>
</li>
<li><code>padj</code>: The p-value adjustment method, see <code>p.adjust.methods</code> in R. Default: <code>BH</code>  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMediation</p>
<ul>
<li>
<p><strong>description</strong><br />
    Do mediation analysis</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file (a matrix or data.frame).  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The result file.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>inopts</code>: The options for input file.  <ul>
<li><code>cnames</code>: Whether the input file has column names</li>
<li><code>rnames</code>: Whether the input file has row names</li>
</ul>
</li>
<li><code>medopts</code>: The options for mediation analysis.  <ul>
<li><code>modelm</code>: The model for M ~ X. Default: <code>lm(M ~ X)</code></li>
<li><code>modely</code>: The model for Y ~ X + M. Default: <code>lm(Y ~ X + M)</code></li>
<li><code>mediator</code>: Tell the model which column is the mediator</li>
<li><code>treat</code>: Tell the model which column is the variable</li>
<li><code>boot</code>: Use bootstrap?</li>
<li><code>sims</code>: How many time simulations?</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pHypergeom</p>
<ul>
<li>
<p><strong>description</strong><br />
    Do hypergeometric test.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file, could be raw data (presence (1) and absence (0) of elements) or number of overlapped elements and elements in each category.  <ul>
<li>Set <code>args.intype</code> as <code>raw</code> if it is raw data. The population size <code>args.N</code> is required</li>
<li>Set <code>args.intype</code> as <code>numbers</code> (or any string except <code>raw</code>) if it is numbers. You can specified explicit header: <code>k</code> = overlapped elements, <code>m</code> = size of set 1, <code>n</code> = size of set 2 and <code>N</code> = the population size. If <code>N</code> not included, then <code>args.N</code> is required</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>intype</code>: the type of input file. Default: <code>raw</code>. See <code>infile:file</code>  </li>
<li><code>inopts</code>: The options for input file.  <ul>
<li><code>cnames</code>: Whether the input file has column names</li>
<li><code>rnames</code>: Whether the input file has row names</li>
</ul>
</li>
<li><code>N</code>: The population size. Default: <code>None</code>  </li>
</ul>
</li>
</ul>
</div>
<h3 id="tabix">tabix</h3>
<div class="admonition hint">
<p class="admonition-title">pTabix</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use tabix to extract information.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile</code>: a local or remote file  </li>
<li><code>region</code>: a region or a file containing regions  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The information extracted from the input file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tabix</code>: The path to <code>tabix</code>  </li>
<li><code>params</code>: Other params for <code>tabix</code>  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pTabixIndex</p>
<ul>
<li>
<p><strong>description</strong><br />
    Generate tabix index file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: the input file  <ul>
<li>Could be bgzipped.</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The bgzipped file  </li>
<li><code>outidx:file</code>: The tabix index file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tabix</code>: The path to <code>tabix</code>  </li>
<li><code>params</code>: Other params for <code>tabix</code>  </li>
<li><code>python</code>: Will be used to generate command line arguments.  </li>
</ul>
</li>
</ul>
</div>
<h3 id="tcga">tcga</h3>
<div class="admonition hint">
<p class="admonition-title">pDownload</p>
<ul>
<li>
<p><strong>description</strong><br />
    Download TCGA use <code>gdc-client</code> and a manifest file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>manifile:file</code>: the manifest file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:file</code>: the directory containing downloaded file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>params</code>: other params for <code>gdc-client</code>, default: "--no-related-files --no-file-md5sum -n 20"  </li>
<li><code>bin-gdc</code>: the executable file of <code>gdc-client</code>, default: "gdc-client"  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSample2SubmitterID</p>
<ul>
<li>
<p><strong>description</strong><br />
    convert TCGA sample names with submitter id with metadata and sample containing folder</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: the directory containing the samples  </li>
<li><code>mdfile:file</code>: the metadata file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:file</code>: the directory containing submitter-id named files  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pConvertExpFiles2Matrix</p>
<ul>
<li>
<p><strong>description</strong><br />
    convert TCGA expression files to expression matrix, and convert sample name to submitter id</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>dir:file</code>: the directory containing the samples  </li>
<li><code>mdfile:file</code>: the metadata file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the output matrix  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://pypi.python.org/pypi/mygene/3.0.0">python-mygene</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pConvertMutFiles2Matrix</p>
<ul>
<li>
<p><strong>description</strong><br />
    convert TCGA mutation files (vcf.gz) to mut matrix, and convert sample name to submitter id</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>dir:file</code>: the directory containing the samples  </li>
<li><code>mdfile:file</code>: the metadata file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the output matrix  </li>
</ul>
</li>
</ul>
</div>
<h3 id="tfbs">tfbs</h3>
<div class="admonition hint">
<p class="admonition-title">pMotifScan</p>
<ul>
<li>
<p><strong>description</strong><br />
    Scan motif along the given sequences.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>tffile:file</code>: The infile containing TF name and motif name.  <ul>
<li>If only one column is give, will be used as both TF and motif name</li>
<li>If there are 2 columns, 1st column will be motif name, 2nd column will be TF name</li>
</ul>
</li>
<li><code>sfile:file</code>: The sequence file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:file</code>: The output dir  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tools</code>   : The tool used to scan the motif. Default: 'meme'  </li>
<li><code>meme</code>    : The path of MEME's fimo. Default: 'fimo'<br />
`motifs   : The motif database in MEME format.</li>
<li><code>pval</code>    : The pvalue cutoff. Default: 1e-4  </li>
<li><code>cleanmname</code>: Whether to clean motif name. Default: True  </li>
<li><code>ucsclink</code>: The ucsc link template. Default: <code>https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&amp;position={}</code>  </li>
<li><code>nthread</code> : Number of threads used to scan, only available when you have multiple mids. Default: 1  </li>
<li><code>params</code>  : Other parameters for <code>fimo</code>  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://meme-suite.org/tools/fimo"><code>fimo</code> from MEME Suite</a></p>
</li>
</ul>
</div>
<h3 id="tsv">tsv</h3>
<div class="admonition hint">
<p class="admonition-title">pMatrixR</p>
<ul>
<li>
<p><strong>description</strong><br />
    Operate a matrix and save the new matrix to file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file containing the matrix  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output matrix  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnames</code>: Whether the input file has cnames. Default: True  </li>
<li><code>rnames</code>: Whether the input file has rnames  . Default: 1  </li>
<li><code>code</code>: The R code to operating the matrix. (the matrix is read in variable <code>mat</code>)  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCbind</p>
<ul>
<li>
<p><strong>description</strong><br />
    Cbind the rest of files to the first file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infiles:files</code>: The input files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output matrix  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnames</code>: Whether the input file has cnames. Default: True  <ul>
<li>or [True, True, False] corresponding to the file order</li>
</ul>
</li>
<li><code>rnames</code>: Whether the input file has rnames  . Default: 1  </li>
<li><code>miss</code>: Replacement for missing values. Default: <code>NA</code>  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pRbind</p>
<ul>
<li>
<p><strong>description</strong><br />
    Rbind the rest of files to the first file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infiles:files</code>: The input files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output matrix  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnames</code>: Whether the input file has cnames. Default: True  <ul>
<li>or [True, True, False] corresponding to the file order</li>
</ul>
</li>
<li><code>rnames</code>: Whether the input file has rnames  . Default: 1  </li>
<li><code>miss</code>: Replacement for missing values. Default: <code>NA</code>  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCsplit</p>
<ul>
<li>
<p><strong>description</strong><br />
    Split a matrix by columns and save them into files.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The directory containing the output column files  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnames</code>: Whether the input file has cnames. Default: True  <ul>
<li>or [True, True, False] corresponding to the file order</li>
</ul>
</li>
<li><code>rnames</code>: Whether the input file has rnames  . Default: 1  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pRsplit</p>
<ul>
<li>
<p><strong>description</strong><br />
    Split a matrix by rows and save them into files.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The directory containing the output row files  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnames</code>: Whether the input file has cnames. Default: True  <ul>
<li>or [True, True, False] corresponding to the file order</li>
</ul>
</li>
<li><code>rnames</code>: Whether the input file has rnames  . Default: 1  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pTsv</p>
<ul>
<li>
<p><strong>description</strong><br />
    Read, Transform, filter a TSV file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>inopts</code>: The input options for infile:  <ul>
<li><code>delimit</code>: The delimit. Default: <code>\\t</code></li>
<li><code>comment</code>: The comment sign. Default: <code>#</code></li>
<li><code>skip</code>: First N lines to skip. Default: <code>0</code></li>
<li><code>ftype</code>: The file type. Metadata can be assigned direct (list/OrderedDict). If not specified, metadata will be generated automatically.</li>
</ul>
</li>
<li><code>outopts</code>: The output options for outfile:  <ul>
<li><code>delimit</code>: The delimit for records. Default: <code>\\t</code></li>
<li><code>head</code>: Output header or not. Default: <code>False</code></li>
<li><code>headDelimit</code>: The delimit for header. Default: <code>\\t</code></li>
<li><code>headPrefix</code>: The prefix for header. Default: ``</li>
<li><code>headTransform</code>: The transformer for header. Default: <code>None</code></li>
<li><code>ftype</code>: The file type. Metadata can be assigned direct (list/OrderedDict, '+' as an element or key is allowed to indicate extra meta from the reader). If not specified, metadata will be borrowed from the reader. </li>
</ul>
</li>
<li><code>ops</code>: A ops function to transform the row. Argument is an instance of <code>readRecord</code>  </li>
<li><code>opshelper</code>: A helper function for <code>args.ops</code>  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSimRead</p>
<ul>
<li>
<p><strong>description</strong><br />
    Read files simultaneously.
    NOTE: only one file allows multiple lines with same value to compare, and that file should be the first one. For example: 
    <div class="highlight"><pre><span></span>File1:
1	1
1	2
1	3
File2:
1	1
2	2
3	3
</pre></div>
    If you compare the first column, File1 has to put at the begining for input.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infiles:files</code>: The input files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>skip</code>: argument skip for each file  </li>
<li><code>delimit</code>: argument delimit for each file  </li>
<li><code>usehead</code>: The header from which input file will be used for output file.  <ul>
<li>Default: None (Don't write header)</li>
</ul>
</li>
<li><code>gzip</code>: argument gzip for each file  </li>
<li><code>match</code>: The match function.   </li>
<li><code>do</code>: The do function. Global vaiable <code>fout</code> is available to write results to output file.  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/pwwang/simread"><code>python-simread</code></a></p>
</li>
</ul>
</div>
<h3 id="utils">utils</h3>
<h3 id="vcf">vcf</h3>
<div class="admonition hint">
<p class="admonition-title">pVcfFilter</p>
<ul>
<li>
<p><strong>description</strong><br />
    Filter records in vcf file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>filters</code>: The filters, should be a string of lambda function:<br />
    <div class="highlight"><pre><span></span>&quot;lambda record, samples: &lt;expression&gt;&quot;
* ``record.CHROM`` : &#39;chr20&#39;
* ``record.POS``   : 1234567
* ``record.ID``    : &#39;microsat1&#39;
* ``record.REF``   : &#39;&#39;GTC&#39;&#39;
* ``record.ALT``   : [G, GTCT]
* ``record.QUAL``  : 50
* ``record.FILTER``: [&#39;PASS&#39;] # NO!, PASS should be []
* ``record.INFO``  : {&#39;AA&#39;: &#39;G&#39;, &#39;NS&#39;: 3, &#39;DP&#39;: 9}
* samples = record.samples
* len(samples): 3
* samples[0].sample: &#39;NA00001&#39;
* samples[0]: Call(sample=NA00001, CallData(GT=0/1, GQ=35, DP=4))
* samples[0].data: calldata(GT=&#39;0/1&#39;, GQ=35, DP=4)
* samples[0].data.GT: &#39;0/1&#39;
</pre></div><ul>
<li>see here for record and samples: https://github.com/jamescasbon/PyVCF</li>
<li>Remember if filters() returns True, record remained.</li>
</ul>
</li>
<li><code>gz</code>     : Whether to gzip the output file. Default: False  </li>
<li><code>keep</code>   : Whether to keep the filtered records. Default: True. (only for gatk, snpsift at filter step)  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/jamescasbon/PyVCF"><code>pyvcf</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pVcf</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use pyvcf to manipulate vcf file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input vcf file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output vcf file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>helper</code>: The helper code injected to script  <ul>
<li>Since lambda function can't do assignment and manipulation so you can write some help function here</li>
</ul>
</li>
<li><code>readerops</code>: A lambda function (must be quoted) to manipulate the reader (vcf.Reader instance)  </li>
<li><code>recordops</code>: A lambda function (must be quoted) to manipulate the record (vcf.Record instance)  </li>
<li><code>gz</code>: Gzip the ouput file  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pVcfAnno</p>
<ul>
<li>
<p><strong>description</strong><br />
    Annotate the variants in vcf file.
    You have to prepare the databases for each tool.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input vcf file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file (output file of annovar will also be converted to vcf)  </li>
<li><code>outdir</code>: The output directory, used to fetch some stat/summary files  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used to do annotation. Default: snpeff  </li>
<li><code>snpeff</code>: The path of snpeff. Default: snpEff  </li>
<li><code>vep</code>: The path to vep. Default: vep  </li>
<li><code>gz</code>: Whether to gzip the result file. Default: False  </li>
<li><code>annovar</code>: The path of annovar. Default: annotate_variation.pl  </li>
<li><code>annovar_convert</code>: The path of convert2annovar.pl, used to convert vcf to annovar input file. Default: convert2annovar.pl  </li>
<li><code>genome</code>: The genome for annotation. Default: hg19  </li>
<li><code>tmpdir</code>: The tmpdir, mainly used by snpeff. Default: <system tmpdir>  </li>
<li><code>dbpath</code>: The path of database for each tool. Required by 'annovar' and 'vep'  </li>
<li><code>params</code>: Other params for tool. Default: ''  </li>
<li><code>snpeffStats</code>: Whether to generate stats file when use snpeff. Default: False  </li>
<li><code>mem</code>: The memory used by snpeff. Default: '4G'  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://doc-openbio.readthedocs.io/projects/annovar/en/latest/"><code>annovar</code></a>
    <a href="http://snpeff.sourceforge.net/SnpEff_manual.html#intro"><code>snpeff</code></a>
    <a href="http://www.ensembl.org/info/docs/tools/vep/script/vep_tutorial.html"><code>vep</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pVcfSplit</p>
<ul>
<li>
<p><strong>description</strong><br />
    Split multi-sample Vcf to single-sample Vcf files.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input vcf file  </li>
<li><code>samples</code>: The samples, if not provided, will extract all samples  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory containing the extracted vcfs  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used to do extraction. Default: vcftools  </li>
<li><code>vcftools</code>: The path of vcftools' vcf-subset  </li>
<li><code>bcftools</code>: The path of bcftools, used to extract the sample names from input vcf file.  </li>
<li><code>gatk</code>: The path of gatk.  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pVcfMerge</p>
<ul>
<li>
<p><strong>description</strong><br />
    Merge single-sample Vcf files to multi-sample Vcf file.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infiles:files</code>: The input vcf files  </li>
<li><code>outfile:dir</code>: The output multi-sample vcf.  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>: The tool used to do extraction. Default: vcftools  </li>
<li><code>vcftools</code>: The path of vcftools' vcf-subset  </li>
<li><code>bcftools</code>: The path of bcftools, used to extract the sample names from input vcf file.  </li>
<li><code>gatk</code>: The path of gatk.  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pVcf2Maf</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert Vcf file to Maf file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code> : The input vcf file  <ul>
<li>see <code>args.somatic</code></li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output maf file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>tool</code>     : Which tool to use. Default: vcf2maf  </li>
<li><code>vcf2maf</code>  : The path of vcf2maf.pl  </li>
<li><code>vep</code>      : The path of vep  </li>
<li><code>vepDb</code>    : The path of database for vep  </li>
<li><code>filtervcf</code>: The filter vcf. Something like: ExAC_nonTCGA.r0.3.1.sites.vep.vcf.gz  </li>
<li><code>ref</code>      : The reference genome  </li>
<li><code>nthread</code>  : Number of threads used to extract samples. Default: 1  </li>
<li><code>tumor1st</code> : Whether tumor sample comes first. Default: <code>True</code>  </li>
<li><code>bcftools</code> : Path to bcftools used to extract sample names.  </li>
<li><code>vcftools</code> : Path to vcftools used to split vcf.  </li>
<li><code>samfunc</code>  : A lambda function used to deduce sample names from file name.  </li>
<li><code>somatic</code>  : Whether input vcf file is a somatic mutation file. Default: False  <ul>
<li>somatic mutation vcf file can only have one sample TUMOR, or two samples, TUMOR and NORMAL, but will be considered as single sample.</li>
<li>otherwise, multiple samples are supported in the input vcf file. Tumor id will be sample name for each sample, normal id will be NORMAL.</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pVcf2GTMat</p>
<ul>
<li><strong>description</strong><br />
    Convert Vcf file to genotype matrix.</li>
</ul>
</div>
<h3 id="vcfnext">vcfnext</h3>
<div class="admonition hint">
<p class="admonition-title">pVcfStatsPlot</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert csvstat file from snpEff to R-readable matrix and plot them.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: The directory containing the csv stat files from <code>snpEff ann</code>  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>chroms</code>: The chromsome filter. Default: "" (all chroms)  </li>
<li>Note: snpEff csvstat file has no "chr" prefix</li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCallRate</p>
<ul>
<li>
<p><strong>description</strong><br />
    Calculate sample/snp call rate from single sample vcfs</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: The dir containing the vcfs  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outsample:file</code>: The report of call rate for each sample  </li>
<li><code>figsample:file</code>: The bar chat of sample call rates  </li>
<li><code>outsnp:file</code>: The report of call rate for each snp  </li>
<li><code>figsnp:file</code>: The bar chat of snp call rates  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCepip</p>
<ul>
<li>
<p><strong>description</strong><br />
    Run CEPIP.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file (vcf or avinput)  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The cepip result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cepip</code>: The path of cepip  </li>
<li><code>cell</code> : The related cell line  </li>
<li><code>params</code>: Other params for cepip  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://jjwanglab.org/cepip/"><code>cepip</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMutSig</p>
<ul>
<li>
<p><strong>description</strong><br />
    MutSig stands for "Mutation Significance".  MutSig analyzes lists of mutations discovered in DNA sequencing, to identify genes that were mutated more often than expected by chance given background mutation processes.
    For more information, see Lawrence, M. et al. Mutational heterogeneity in cancer and the search for new cancer-associated genes. Nature 499, 214-218 (2013).</p>
<p>See <a href="http://archive.broadinstitute.org/cancer/cga/mutsig_run">dcumentation</a></p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: mutation table  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>mutsig</code> : The path to <code>run_MutSigCV.sh</code>, default: 'mutsig'  </li>
<li><code>mcr</code>    : The Matlab MCR path  </li>
<li><code>cvrg</code>   : coverage table  </li>
<li><code>cvrt</code>   : covariates table  </li>
<li><code>mutdict</code>: mutation_type_dictionary_file  </li>
<li><code>chrdir</code> : chr_files_hg18 or chr_files_hg19  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://archive.broadinstitute.org/cancer/cga/mutsig_download">MutSig</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMafMerge</p>
<ul>
<li>
<p><strong>description</strong><br />
    Merge maf files.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:dir</code>: The directory containing the maf files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The merged maf file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>excols</code>: How to deal with extra columns other than 34 standard columns from TCGA.  <ul>
<li>merge(default): Merge the columns, if one not exists, fill with an empty string.</li>
<li>discard: Just discard the extra columns, with only 34 columns left. So you can also put just one maf file in the indir with some columns missed to fill it with standard columns.</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMaftools</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use maftools to draw plots.</p>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>ngenes</code>:   </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
    [``]</p>
</li>
</ul>
</div>
<h3 id="web">web</h3>
<div class="admonition hint">
<p class="admonition-title">pDownloadForm</p>
<ul>
<li>
<p><strong>description</strong><br />
    Download results by submitting a form, supporting pagination.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>url</code>   : the URL contains the form  </li>
<li><code>data</code>  : the data used to fill the form (JSON string or transformed from dict by json.dumps).  </li>
<li><code>submit</code>: the submit button to submit the form (use Xpath).  </li>
<li><code>next</code>  : the button for next page (use Xpath)  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:file</code>: The directory saves the results  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>interval</code>: seconds to wait between fetching each page. Default: 1  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://splinter.readthedocs.io/en/latest/index.html"><code>Splinter</code></a>
    <a href="http://phantomjs.org/"><code>Phantomjs</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pDownloadGet</p>
<ul>
<li>
<p><strong>description</strong><br />
    Download results by urls.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>url</code>: the URLs to download  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pDownload</p>
<ul>
<li><strong>description</strong><br />
    Alias of <code>pDownloadGet</code></li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pDownloadPost</p>
<ul>
<li>
<p><strong>description</strong><br />
    Download results by POST.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>url</code> : the URLs to download  </li>
<li><code>data</code>: the POST data.  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
</ul>
</div>
<h3 id="wxs">wxs</h3>
<h3 id="wxsanno">wxsanno</h3>
<div class="admonition hint">
<p class="admonition-title">pSnpEff</p>
<ul>
<li>
<p><strong>description</strong><br />
    This is the default command. It is used for annotating variant filed (e.g. VCF files).</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The input file   </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:file</code>: The directory containing output anntated file, snpEff_genes.txt and snpEff_summary.html  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>snpEff</code>: The snpEff executable, default: "snpEff"  </li>
<li><code>params</code>: Other parameters for <code>snpEff</code>, default: "-Xms1g -Xmx4g -v"  </li>
<li><code>genome</code>: The genome used for annotation, default: "hg19"  </li>
<li><code>informat</code>: The format of input file [vcf or bed], default: "vcf"  </li>
<li><code>outformat</code>: The format of output file [vcf, gatk, bed, bedAnn], default: "vcf"  </li>
<li><code>csvStats</code>: Whether to generate csv stats file, default: True.  </li>
<li><code>htmlStats</code>: Whether to generate the html summary file, default: False.  </li>
<li><code>javamem</code>: The memory to use. Default: '-Xms1g -Xmx8g'  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://snpeff.sourceforge.net/SnpEff_manual.html">snpEff</a></p>
</li>
</ul>
</div>
<h3 id="wxscall">wxscall</h3>
<div class="admonition hint">
<p class="admonition-title">pCNVnator</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use <code>CNVnator</code> to call CNVs from bam file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The bam file   </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The vcf file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>cnvnator</code>: The CNVnator executable, default: "cnvnator"  </li>
<li><code>cnv2vcf</code>: The converter executable to convert CNVnator results to vcf, default: "cnvnator2VCF.pl"  </li>
<li><code>binsize</code>: The bin_size, default: 100  </li>
<li><code>genome</code>: The genome: default: hg19  </li>
<li><code>chrom</code>: Chromosome names, default: "" (all chromosomes)  </li>
<li><code>chrdir</code>: The dir contains reference sequence of chromosomes, default: "" (don't specify)  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/abyzovlab/CNVnator">CNVnator</a></p>
</li>
</ul>
</div>
<h3 id="wxsdown">wxsdown</h3>
<div class="admonition hint">
<p class="admonition-title">pMutSig</p>
<ul>
<li>
<p><strong>description</strong><br />
    MutSig stands for "Mutation Significance".  MutSig analyzes lists of mutations discovered in DNA sequencing, to identify genes that were mutated more often than expected by chance given background mutation processes.</p>
<p>For more information, see Lawrence, M. et al. Mutational heterogeneity in cancer and the search for new cancer-associated genes. Nature 499, 214-218 (2013).</p>
<p>See <a href="http://archive.broadinstitute.org/cancer/cga/mutsig_run">dcumentation</a></p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>maffile:file</code>: mutation table  </li>
<li><code>cvgfile:file</code>: coverage table  </li>
<li><code>cvrfile:file</code>: covariates table  </li>
<li><code>mutdict:file</code>: mutation_type_dictionary_file  </li>
<li><code>chrdir:file</code>: chr_files_hg18 or chr_files_hg19   </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>mutsig</code>: The path to <code>run_MutSigCV.sh</code>, default: 'mutsig'  </li>
<li><code>mcr</code>: The Matlab MCR path  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://archive.broadinstitute.org/cancer/cga/mutsig_download">MutSing</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pVcf2Maf</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert a snpEff-annotated somatic mutation vcf file (with normal and tumor samples) to <a href="https://wiki.nci.nih.gov/display/TCGA/Mutation+Annotation+Format+(MAF)+Specification">maf</a> file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: vcf file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The maf file  </li>
</ul>
</li>
<li>
<p><strong>args</strong><br />
<code>vepdata</code>: The path of vep data. Default: "" (default data dir of vep)
       <code>vep</code>: The path of vep excutable. Default: "vep"
       <code>vcf2maf</code>: The path of vcf2maf excutable. Default: "vcf2maf.pl"
       <code>reffile</code>: The reference fasta file.
       <code>nthread</code>: The number of threads used by vep. Default: 1
       <code>filtervcf</code>: The filter vcf</p>
<ul>
<li><code>params</code>: Other parameters for <code>vcf2maf.pl</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/mskcc/vcf2maf">vcf2maf.py</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMergeMafs</p>
<ul>
<li>
<p><strong>description</strong><br />
    Merge MAF files</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: The directory containing MAF files to be merged  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The merged MAF file  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMutsig4Plot</p>
<ul>
<li>
<p><strong>description</strong><br />
    Prepare somatic mutations for  plotting</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>msdir:file</code>: The mutsig output directory  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The file for plotting<br />
<div class="highlight"><pre><span></span>#PANEL: Somatic mutations
#INFO: MT|PI
#DESC: Mutation type|Putative impact
# could also be bordercolor, you can have up to 4 shape features
#TYPE: shape|bgcolor
# could also be continuous
# expressions for set: a,b,c
#                 norminal: no
#                 continuous: [0,1]
#DATA: set|norminal
#NCOL: 2|2
#NAME_MT: Frameshift|Missense|Nonsense|Silent|Splice_site|TSS|Nonstop
#NAME_PI: HIGH|MODERATE|LOW|MODIFIER
#VALUE_MT: 0|1|20|13|4|17|14
#EXP_MT: frameshift_variant,inframe_deletion,inframe_insertion|missense_variant,initiator_codon_variant,stop_retained_variant,rare_amino_acid_variant|stop_gained|synonymous_variant|splice_acceptor_variant,splice_donor_variant|start_lost,start_retained|stop_lost
#
Sample1	Sample2	Sample3	Sample4	Sample5
ABC	missense_variant|HIGH	missense_variant|HIGH	...
...
</pre></div></li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>topn</code>: the cutoff to select genes. If it is &gt;= 1, top N genes will be selected, otherwise, it will be used as pvalue cutoff. Default: .05  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/jamescasbon/PyVCF"><code>pyvcf</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMutPlot</p>
<ul>
<li>
<p><strong>description</strong><br />
    Plot mutations
    <div class="highlight"><pre><span></span>|           |             |           |           |---
|- ftWidth -|  s   s   s  |- pnWidth -|- lgWidth -| snHeight
|           |             |           |           |---
    feature1
	feature2
</pre></div></p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: The input directory containing plot files  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The plot png file  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCepip</p>
<ul>
<li>
<p><strong>description</strong><br />
    run CEPIP.</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>avinput:file</code>: The avinput file  </li>
<li><code>cell</code>: The cell  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The cepip result file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bin-cepip</code>: The jar file path of cepip, default: /data2/junwenwang/shared/tools/cepip/cepip.jar  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://jjwanglab.org/cepip/"><code>cepip</code></a></p>
</li>
</ul>
</div>
<h3 id="wxsprep">wxsprep</h3>
<div class="admonition hint">
<p class="admonition-title">pTrimmomaticPE</p>
<ul>
<li>
<p><strong>description</strong><br />
    Trimming Illumina NGS paired-end data</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>fqfile1:file</code>: The 1st fastq file (could be in .gz format)  </li>
<li><code>fqfile2:file</code>: The 2nd fastq file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile1:file</code>: The 1st output file  </li>
<li><code>outfile2:file</code>: The 2nd output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>trimmomatic</code>: The trimmomatic executable, default: "trimmomatic"  </li>
<li><code>phred</code>: "phred33" (default) or "phred64"  </li>
<li><code>params</code>: Other params for trimmomatric, default: "ILLUMINACLIP:{adapter}:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36"  <ul>
<li>have to replace <code>{adapter}</code> with the path of the adapter file</li>
</ul>
</li>
<li><code>nthread</code>: 1  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://www.usadellab.org/cms/index.php?page=trimmomatic">trimmomatic</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pTrimmomaticSE</p>
<ul>
<li>
<p><strong>description</strong><br />
    Trimming Illumina NGS single-end data</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>fqfile:file</code>: The fastq file (could be in .gz format)  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>trimmomatic</code>: The trimmomatic executable, default: "trimmomatic"  </li>
<li><code>phred</code>: "phred33" (default) or "phred64"  </li>
<li><code>params</code>: Other params for trimmomatric, default: "ILLUMINACLIP:{adapter}:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36"  <ul>
<li>have to replace <code>{adapter}</code> with the path of the adapter file</li>
</ul>
</li>
<li><code>nthread</code>: 1  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://www.usadellab.org/cms/index.php?page=trimmomatic">trimmomatic</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pAlignPEByBWA</p>
<ul>
<li>
<p><strong>description</strong><br />
    Align paired-end reads to reference genome using bwa mem</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile1:file</code>: read file 1 (fastq, or fastq gzipped)  </li>
<li><code>infile2:file</code>: read file 2 (fastq, or fastq gzipped)  </li>
<li><code>reffile:file</code>: The reference file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output sam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bwa</code>: The bwa executable, default: bwa  </li>
<li><code>params</code>: Other params for bwa mem, default: "-M"  </li>
<li><code>nthread</code>: 1  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/lh3/bwa">bwa</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pAlignSEByBWA</p>
<ul>
<li>
<p><strong>description</strong><br />
    Align paired-end reads to reference genome using bwa mem</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: read file (fastq, or fastq gzipped)  </li>
<li><code>reffile:file</code>: The reference file  </li>
</ul>
</li>
<li>
<p><strong>brings</strong>  </p>
<ul>
<li><code>reffile#bwt</code>: "{{reffile | bn}}.bwt",   </li>
<li><code>reffile#sa</code>: "{{reffile | bn}}.sa",  </li>
<li><code>reffile#ann</code>: "{{reffile | bn}}.ann",  </li>
<li><code>reffile#amb</code>: "{{reffile | bn}}.amb",  </li>
<li><code>reffile#pac</code>: "{{reffile | bn}}.pac"  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output sam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bwa</code>: The bwa executable, default: bwa  </li>
<li><code>params</code>: Other params for bwa mem, default: "-M"  </li>
<li><code>nthread</code>: 1  </li>
<li><code>reffile</code>: The reference file, required  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/lh3/bwa">bwa</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pAlignPEByNGM</p>
<ul>
<li>
<p><strong>description</strong><br />
    Align paired-end reads to reference genome using NextGenMap</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile1:file</code>: read file 1 (fastq, or fastq gzipped)  </li>
<li><code>infile2:file</code>: read file 2 (fastq, or fastq gzipped)  </li>
<li><code>reffile:file</code>: The reference file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output sam/bam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>ngm</code>: The NextGenMap executable, default: ngm  </li>
<li><code>nthread</code>: 1  </li>
<li><code>outtype</code>: sam or bam, default: sam (only sam for now, due to bug of ngm 0.5.3 (fixed in 0.5.4))  </li>
<li><code>params</code>: Other params for ngm, default: "--rg-id ngm --rg-sm sample"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/Cibiv/NextGenMap/wiki">NextGenMap</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pAlignSEByNGM</p>
<ul>
<li>
<p><strong>description</strong><br />
    Align single-end reads to reference genome using NextGenMap</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile1:file</code>: read file 1 (fastq, or fastq gzipped)  </li>
<li><code>infile2:file</code>: read file 2 (fastq, or fastq gzipped)  </li>
<li><code>reffile:file</code>: The reference file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The output sam/bam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>ngm</code>: The NextGenMap executable, default: ngm  </li>
<li><code>nthread</code>: 1  </li>
<li><code>outtype</code>: sam or bam, default: sam (only sam for now, due to bug of ngm 0.5.3 (fixed in 0.5.4))  </li>
<li><code>params</code>: Other params for ngm, default: "--rg-id ngm --rg-sm sample"  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/Cibiv/NextGenMap/wiki">NextGenMap</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pMergeBams</p>
<ul>
<li>
<p><strong>description</strong><br />
    Merge bam files</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>bamdir:dir</code>: the dir containing bam files   </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: the merged bam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>samtools</code>: the executable path of samtools, default: "samtools"  </li>
<li><code>nthread</code>: Number of BAM/CRAM compression threads  </li>
<li><code>params</code>: Other parameters for <code>samtools merge</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://www.htslib.org/">samtools</a></p>
</li>
</ul>
</div>
<h3 id="wxsstat">wxsstat</h3>
<div class="admonition hint">
<p class="admonition-title">pVcf2List</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert vcf to stat files for pCallRate</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>vcffile:file</code>: The vcf file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The stat file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>chroms</code>: SNPs on chromosomes to consider, default: "" (all chroms)  </li>
<li>use "chr1-22, chrX, chrY" for chr1 to chr22, chrX and chrY</li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/jamescasbon/PyVCF"><code>pyvcf</code></a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCallRate</p>
<ul>
<li>
<p><strong>description</strong><br />
    Calculate sample/snp call rate from a matrix of snp-sample</p>
<ul>
<li>rows are snps, columns are samples</li>
</ul>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The snp-sample matrix file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outsample:file</code>: The report of call rate for each sample  </li>
<li><code>figsample:file</code>: The bar chat of sample call rates  </li>
<li><code>outsnp:file</code>: The report of call rate for each snp  </li>
<li><code>figsnp:file</code>: The bar chat of snp call rates  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pCoverageByBamstats</p>
<ul>
<li>
<p><strong>description</strong><br />
    Use <code>bamstats</code> to calculate coverage for bam file</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>infile:file</code>: The bam file  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outfile:file</code>: The report of coverage for the bam file  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>bin</code>: The <code>bamstats</code> executable, default: "bamstats"  </li>
<li><code>params</code>: Other parameters for <code>bamstats</code>, default: ""  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="http://bamstats.sourceforge.net/">bamstats</a></p>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pPlotBamstats</p>
<ul>
<li>
<p><strong>description</strong><br />
    Plot coverage use output files generated by <code>bamstats</code> or <code>wxs.pCoverageByBamstats</code></p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: The directory containing bamstats output files  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>chroms</code>: Chromosomes to plot. Default: "" (all chroms)  </li>
<li>Note: Whether to have "chr" prefix or not depends on your reference when mapping.</li>
<li>You can do a scope assignment: "chr1-chr22, chrX, chrY"</li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:file</code>: The directory containing output figures  </li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pSnpEff2Stat</p>
<ul>
<li>
<p><strong>description</strong><br />
    Convert csvstat file from snpEff to R-readable matrix for plotting</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: The directory containing the csv stat files from <code>snpEff ann</code>  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory  </li>
</ul>
</li>
<li>
<p><strong>args</strong>  </p>
<ul>
<li><code>chroms</code>: The chromsome filter. Default: "" (all chroms)  </li>
<li>Note: snpEff csvstat file has no "chr" prefix</li>
</ul>
</li>
</ul>
</div>
<div class="admonition hint">
<p class="admonition-title">pPlotSnpEff</p>
<ul>
<li>
<p><strong>description</strong><br />
    Plot snpEff annotation statistics</p>
</li>
<li>
<p><strong>input</strong>  </p>
<ul>
<li><code>indir:file</code>: The snpEff result directory containing matrix files generated by pSnpEff2Stat  </li>
</ul>
</li>
<li>
<p><strong>output</strong>  </p>
<ul>
<li><code>outdir:dir</code>: The output directory  </li>
</ul>
</li>
<li>
<p><strong>requires</strong><br />
<a href="https://github.com/pwwang/corrplot"><code>pwwang/corrplot</code></a></p>
<ul>
<li>use <code>library(devtools); install.github("pwwang/corrplot")</code>
<a href="http://ggplot2.org/"><code>ggplot2</code></a></li>
</ul>
</li>
</ul>
</div>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../aggregations/" class="btn btn-neutral float-right" title="Aggregations">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href=".." class="btn btn-neutral" title="Introduction"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href=".." style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../aggregations/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js"></script>
      <script src="../search/require.js"></script>
      <script src="../search/search.js"></script>

</body>
</html>
