{% raw %}

## pExpdir2Matrix

### description
	Convert expression files to expression matrix
	File names will be used as sample names (colnames)
	Each gene and its expression per line.
	Suppose each expression file has the same rownames and in the same order.

### input
#### `expdir:file`:
  the directory containing the expression files, could be gzipped  

### output
#### `outfile:file`:
 the expression matrix file  
#### `outdir:dir`:
   the directory containing expr file and plots  

## pBatchEffect

### description
	Remove batch effect with sva-combat.

### input
#### `expr:file`:
  The expression file, generated by pExpdir2Matrix  
#### `batch:file`:
 The batch file defines samples and batches.  

### output
#### `outfile:file`:
 the expression matrix file  
#### `outdir:dir`:
   the directory containing expr file and plots  

## pRawCounts2

### description
	Convert raw counts to another unit

### input
#### `expfile:file`:
 the expression matrix  
		- rows are genes, columns are samples

### output
#### `outfile:file`:
 the converted expression matrix  

### args
#### `transpose`:
 transpose the input matrix? default: False  
#### `log2`:
      whether to take log2? default: False  
#### `unit`:
      convert to which unit? default: cpm (or rpkm, tmm)  
#### `header`:
    whether input file has header? default: True  
#### `rownames`:
  the index of the column as rownames. default: 1  
#### `glenfile`:
  the gene length file, for RPKM  
		- no head, row names are genes, have to be exact the same order and length as the rownames of expfile
#### `boxplot` :
 Whether to plot a boxplot. Default: False  
#### `heatmap` :
 Whether to plot a heatmap. Default: False  
#### `histplot`:
 Whether to plot a histgram. Default: False  
#### `devpars` :
 Parameters for png. Default: `{'res': 300, 'width': 2000, 'height': 2000}`  
#### `boxplotggs`:
 The ggplot parameters for boxplot. Default: `['r:ylab("Expression")']`  
		- See ggplot2 documentation.
#### `heatmapggs`:
 The ggplot parameters for heatmap. Default: `['r:theme(axis.text.y = element_blank())']`  
#### `histplotggs`:
 The ggplot parameters for histgram. Default: `['r:labs(x = "Expression", y = "# Samples")']`	  

## p2RawCounts

### description
	Convert gene expression to raw counts.

### input
#### `expfile:file`:
 The expression matrix file  

### output
#### `outfile:file`:
 The output file  
#### `outdir:dir`:
   The output directory, may contain the figures.  

## pRnaseqDeg

### description
	Detect DEGs for RNA-seq data

### input
#### `efile:file`:
 The expression matrix  
#### `gfile:file`:
 The group information  
		- Like:
		```
		Sample1	Group1
		Sample2	Group1
		Sample3	Group1
		Sample4	group2
		Sample5	group2
		Sample6	group2
		```

### output
#### `outfile:file`:
 The DEG list  
#### `outdir:file`:
  The output directory containing deg list and plots  
{% endraw %}
