# snparray
<!-- toc -->
{% raw %}

## pGistic

### description
Runing GISTIC to get CNV results.
see: ftp://ftp.broadinstitute.org/pub/GISTIC2.0/GISTICDocumentation_standalone.htm

### input
#### `segfile:file`:: Segmentation File  
#### `mkfile:file` :: Markers File  
#### `alfile:file` :: Array List File  
#### `cnvfile:file`:: CNV File  

### output
#### `outdir:dir`:: The output directory  
	- All Lesions File (all_lesions.conf_XX.txt, where XX is the confidence level)
	- Amplification Genes File (amp_genes.conf_XX.txt, where XX is the confidence level)
	- Deletion Genes File (del_genes.conf_XX.txt, where XX is the confidence level)
	- Gistic Scores File (scores.gistic)
	- Segmented Copy Number (raw_copy_number.pdf)

### args
#### `gistic`:: The path to gistic.  
#### `genome`:: The genome used to select refgene file from refgenefiles.  
#### `mcr`::    The mcr path  
#### `params`:: Other params for gistic  

## pSNP6Genotype

### description
Call genotypes from GenomeWideSNP_6 CEL file

### input
#### `celfile:file`:: the CEL file  

### output
#### `outfile:file`:: the outfile containing probe name and genotypes  
- format: `<Probe name>\t<genotype>`
- `<genotype>` = 0: AA, 1: AB, 2: BB

### requires
[bioconductor-crlmm](http://bioconductor.org/packages/release/bioc/html/crlmm.html)

## pGenoToAvInput

### description
Convert the genotype called by pSNP6Genotype to [ANNOVAR input file](http://annovar.openbioinformatics.org/en/latest/user-guide/input/#annovar-input-file) using dbSNP identifiers.	

### input
#### `genofile:file`:: the genofile generated by pSNP6Genotype, must be sorted by probe names  
#### `annofile:flie`:: the annotation file downloaded from http://www.affymetrix.com/support/technical/annotationfilesmain.affx  
	- Could be in .gz format

### output
#### `outfile:file`:: the avinput file  

### requires
[python-read2](https://github.com/pwwang/read2)
{% endraw %}
