# wxsstat
<!-- toc -->
{% raw %}

## pVcf2List

### description
Convert vcf to stat files for pCallRate

### input
#### `vcffile:file`:: The vcf file  

### output
#### `outfile:file`:: The stat file  

### args
#### `chroms`:: SNPs on chromosomes to consider, default: "" (all chroms)  
- use "chr1-22, chrX, chrY" for chr1 to chr22, chrX and chrY

### requires
[`pyvcf`](https://github.com/jamescasbon/PyVCF)

## pCallRate

### description
Calculate sample/snp call rate from a matrix of snp-sample
- rows are snps, columns are samples

### input
#### `infile:file`::    The snp-sample matrix file  

### output
#### `outsample:file`:: The report of call rate for each sample  
#### `figsample:file`:: The bar chat of sample call rates  
#### `outsnp:file`::    The report of call rate for each snp  
#### `figsnp:file`::    The bar chat of snp call rates  

## pCoverageByBamstats

### description
Use `bamstats` to calculate coverage for bam file

### input
#### `infile:file`::  The bam file  

### output
#### `outfile:file`::    The report of coverage for the bam file  

### args
#### `bin`:: The `bamstats` executable, default: "bamstats"  
#### `params`:: Other parameters for `bamstats`, default: ""  

### requires
[bamstats](http://bamstats.sourceforge.net/)

## pPlotBamstats

### description
Plot coverage use output files generated by `bamstats` or `wxs.pCoverageByBamstats`

### input
#### `indir:file`:: The directory containing bamstats output files  

### args
#### `chroms`:: Chromosomes to plot. Default: "" (all chroms)  
- Note: Whether to have "chr" prefix or not depends on your reference when mapping.
- You can do a scope assignment: "chr1-chr22, chrX, chrY"

### output
#### `outdir:file`:: The directory containing output figures  

## pSnpEff2Stat

### description
Convert csvstat file from snpEff to R-readable matrix for plotting

### input
#### `indir:file`:: The directory containing the csv stat files from `snpEff ann`  

### output
#### `outdir:dir`:: The output directory  

### args
#### `chroms`::     The chromsome filter. Default: "" (all chroms)  
- Note: snpEff csvstat file has no "chr" prefix

## pPlotSnpEff

### description
Plot snpEff annotation statistics

### input
#### `indir:file`:: The snpEff result directory containing matrix files generated by pSnpEff2Stat  

### output
#### `outdir:dir`:: The output directory  

### requires
[`pwwang/corrplot`](https://github.com/pwwang/corrplot)
- use `library(devtools); install.github("pwwang/corrplot")`
[`ggplot2`](http://ggplot2.org/)
{% endraw %}
